{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Paper next results\n",
    "- validate link between arrhythmia, myopathy, and the identified genes (SLCO1B1)\n",
    "- validate the drugs involved in arrhythmia\n",
    "    - run PoSe-Path on the drugs found in literature for arrhythmia; check for other side effects (and check in literature)\n",
    "- validate the other promising heart side effects\n",
    "    - explain (jointly) related side effects (having similar mechanisms) in the heart and across organs (e.g., heart muscle, other muscle\n",
    "- validate muscle (myopathy) since connected with cardiovascular disease\n",
    "- move on to validating liver (since SLCO1B1 is in liver), kidney, (brain)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Explainer paper\n",
    "1. Each single side effect --> Enriched GO --> validate with literature\n",
    "1. Choose two side effects suspected to share mechanisms --> get common subgraph --> validate\n",
    "1. Choose >2 side effects ...\n",
    "1. Choose single interesting drug --> Enriched GO --> validate with literature\n",
    "1. Choose interesting drug-drug interaction --> Enriched GO --> validate with literature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import goatools\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import pubchempy\n",
    "import pickle\n",
    "\n",
    "from pubchempy import Compound\n",
    "\n",
    "# Get http://geneontology.org/ontology/go-basic.obo\n",
    "from goatools.base import download_go_basic_obo\n",
    "# Get ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene2go.gz\n",
    "from goatools.base import download_ncbi_associations\n",
    "from goatools.obo_parser import GODag\n",
    "from __future__ import print_function\n",
    "from goatools.anno.genetogo_reader import Gene2GoReader\n",
    "from goatools.test_data.genes_NCBI_9606_ProteinCoding import GENEID2NT as GeneID2nt_hum\n",
    "from goatools.goea.go_enrichment_ns import GOEnrichmentStudyNS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/laurence/git/posepath/PoSe-Path-3.0/data/tipexp_data.pkl','rb') as f:\n",
    "    mapping = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "536"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping.side_effect_name_to_idx['Arrhythmia']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{2: 'atelectasis',\n",
       " 4: 'Back Ache',\n",
       " 5: 'lung edema',\n",
       " 6: 'agitated',\n",
       " 7: 'abnormal movements',\n",
       " 8: 'Acidosis',\n",
       " 9: 'peliosis',\n",
       " 11: 'Apnea',\n",
       " 12: 'Drug hypersensitivity',\n",
       " 13: 'flatulence',\n",
       " 14: 'pain in throat',\n",
       " 15: 'allergies',\n",
       " 16: 'thrombocytopenia',\n",
       " 17: 'bradycardia',\n",
       " 18: 'lung infiltration',\n",
       " 19: 'Bleeding',\n",
       " 20: 'hypoglycaemia neonatal',\n",
       " 22: 'hyperglycaemia',\n",
       " 23: 'peritonitis',\n",
       " 24: 'hypoglycaemia',\n",
       " 25: 'abdominal distension',\n",
       " 26: 'asystole',\n",
       " 27: 'cerebral infarct',\n",
       " 28: 'hypoxia',\n",
       " 29: 'Difficulty breathing',\n",
       " 30: 'decreased body temperature',\n",
       " 31: 'adynamic ileus',\n",
       " 32: 'sepsis',\n",
       " 33: 'Hypomagnesaemia',\n",
       " 35: 'Acute Respiratory Distress Syndrome',\n",
       " 36: 'lipoma',\n",
       " 38: 'appendectomy',\n",
       " 39: 'flank pain',\n",
       " 40: 'pneumothorax',\n",
       " 41: 'appendicitis',\n",
       " 42: 'Strabismus',\n",
       " 43: 'Blood calcium decreased',\n",
       " 44: 'respiratory failure',\n",
       " 45: 'pulmonary hypertension',\n",
       " 46: 'Hypercapnia',\n",
       " 47: 'intestinal perforation',\n",
       " 48: 'leucocytosis',\n",
       " 49: 'Disorder Lung',\n",
       " 50: 'hot flash',\n",
       " 51: 'arterial pressure NOS decreased',\n",
       " 52: 'metabolic acidosis',\n",
       " 53: 'Bladder inflammation',\n",
       " 54: 'ascites',\n",
       " 56: 'sinus tachycardia',\n",
       " 57: 'stridor',\n",
       " 58: 'arterial pressure NOS increased',\n",
       " 59: 'Hypoventilation',\n",
       " 60: 'edema',\n",
       " 61: 'disease of liver',\n",
       " 62: 'Aspartate Aminotransferase Increase',\n",
       " 63: 'Cold extremities',\n",
       " 64: 'facial flushing',\n",
       " 65: 'anaphylactic reaction',\n",
       " 66: 'eruption',\n",
       " 67: 'acute pancreatitis',\n",
       " 68: 'neumonia',\n",
       " 69: 'adenocarcinoma',\n",
       " 70: 'confusion',\n",
       " 71: 'malnourished',\n",
       " 72: 'haemorrhoids',\n",
       " 73: 'still birth',\n",
       " 74: 'anaemia',\n",
       " 75: 'drug toxicity NOS',\n",
       " 76: 'increased white blood cell count',\n",
       " 77: 'encephalopathy',\n",
       " 78: 'Anorexia',\n",
       " 79: 'tremor',\n",
       " 80: 'Hallucination',\n",
       " 81: 'chronic obstructive airway disease',\n",
       " 82: 'drowsiness',\n",
       " 83: 'Head ache',\n",
       " 84: 'back injury',\n",
       " 85: 'erythema',\n",
       " 86: 'Diplopia',\n",
       " 87: 'enlarged spleen',\n",
       " 88: 'narcolepsy',\n",
       " 89: 'esophageal spasm',\n",
       " 90: 'erosive gastritis',\n",
       " 91: 'gastroenteritis viral',\n",
       " 92: 'Lymphocytes decreased',\n",
       " 93: 'Breast cancer',\n",
       " 94: 'itch',\n",
       " 95: 'haemolytic anaemia',\n",
       " 96: 'upper gastrointestinal bleeding',\n",
       " 97: 'icterus',\n",
       " 98: 'gastroenteritis',\n",
       " 99: 'optic disc edema',\n",
       " 100: 'sleep disorder',\n",
       " 101: 'septic shock',\n",
       " 102: 'bladder retention',\n",
       " 103: 'Aphonia',\n",
       " 104: 'Pain',\n",
       " 105: 'idiopathic thrombocytopenic purpura',\n",
       " 106: 'pneumonia staphylococcal',\n",
       " 107: 'alopecia',\n",
       " 108: 'conjunctivitis',\n",
       " 109: 'lobar pneumonia',\n",
       " 110: 'glomerulonephritis',\n",
       " 111: 'duodenal ulcer',\n",
       " 112: 'deep vein thromboses',\n",
       " 113: 'leucopenia',\n",
       " 114: 'rhinorrhea',\n",
       " 115: 'bursitis',\n",
       " 116: 'bleeding Vaginal',\n",
       " 117: 'Excoriation',\n",
       " 118: 'Ekbom Syndrome',\n",
       " 119: 'nephrocalcinosis',\n",
       " 120: 'blurred vision',\n",
       " 121: 'thrombotic microangiopathy',\n",
       " 122: 'dizziness',\n",
       " 123: 'pleurisy',\n",
       " 124: 'sunburn',\n",
       " 125: 'Hoarseness',\n",
       " 126: 'Bleeding hemorrhoids',\n",
       " 127: 'Fatigue',\n",
       " 128: 'aspiration pneumonia',\n",
       " 129: 'perirectal abscess',\n",
       " 130: 'diarrhea',\n",
       " 131: 'asthma',\n",
       " 132: 'skin lesion',\n",
       " 133: 'tinnitus',\n",
       " 134: 'muscle spasm',\n",
       " 135: 'fib',\n",
       " 136: 'Candida Infection',\n",
       " 137: 'Abnormal Gait',\n",
       " 138: 'glossodynia',\n",
       " 139: 'anal fissure',\n",
       " 140: 'Infection Viral',\n",
       " 141: 'cellulitis',\n",
       " 142: 'allergic dermatitis',\n",
       " 143: 'constipated',\n",
       " 144: 'Clotting',\n",
       " 145: 'muscle paresis',\n",
       " 146: 'deafness',\n",
       " 147: 'vocal cord paralysis',\n",
       " 148: 'dry skin',\n",
       " 150: 'eye pain',\n",
       " 151: 'Leukaemia',\n",
       " 153: 'Chronic Kidney Disease',\n",
       " 154: 'acute kidney failure',\n",
       " 155: 'retrocollis',\n",
       " 156: 'cognitive disorder',\n",
       " 157: 'Cholangitis',\n",
       " 158: 'hypothyroid',\n",
       " 159: 'dermatitides',\n",
       " 160: 'Abnormal LFTs',\n",
       " 161: 'dermatitis medicamentosa',\n",
       " 162: 'acute brain syndrome',\n",
       " 163: 'Deafness neurosensory',\n",
       " 164: 'fungal disease',\n",
       " 165: 'nasal sinus congestion',\n",
       " 166: 'Near Syncope',\n",
       " 167: 'erythema multiforme',\n",
       " 168: 'haematoma',\n",
       " 169: 'bleb',\n",
       " 170: 'nasal congestion',\n",
       " 171: 'amenorrhea',\n",
       " 172: 'epistaxis',\n",
       " 173: 'hepatitis',\n",
       " 174: 'coated tongue',\n",
       " 175: 'Neutropenia',\n",
       " 176: 'wheeze',\n",
       " 177: 'haemorrhage rectum',\n",
       " 178: 'abdominal pain',\n",
       " 179: 'chest pain',\n",
       " 180: 'Mycobacterium tuberculosis infection',\n",
       " 181: 'cholelithiasis',\n",
       " 182: 'pyoderma',\n",
       " 184: 'Blood disorder',\n",
       " 185: 'psychoses',\n",
       " 186: 'aching muscles',\n",
       " 187: 'abscess',\n",
       " 188: 'loss of consciousness',\n",
       " 189: 'attempted suicide',\n",
       " 190: 'bad breath',\n",
       " 191: 'Arthritis infective',\n",
       " 192: 'loss of weight',\n",
       " 193: 'nasopharyngitis',\n",
       " 194: 'belching',\n",
       " 195: 'Rhinitis',\n",
       " 196: 'Colon Spastic',\n",
       " 197: 'heart attack',\n",
       " 198: 'hepatic neoplasia',\n",
       " 199: 'azotaemia',\n",
       " 200: 'nervous tension',\n",
       " 201: 'muscle weakness',\n",
       " 202: 'hypochloraemia',\n",
       " 203: 'cataract',\n",
       " 204: 'abdominal pain upper',\n",
       " 205: 'insomnia',\n",
       " 206: 'dyspepsia',\n",
       " 207: 'asthenia',\n",
       " 208: 'emesis',\n",
       " 209: 'Embolism pulmonary',\n",
       " 210: 'biliary tract disorder',\n",
       " 211: 'colitis ischemic',\n",
       " 212: 'Extremity pain',\n",
       " 213: 'arteriosclerotic heart disease',\n",
       " 214: 'Small intestinal obstruction',\n",
       " 215: 'Paraesthesia',\n",
       " 216: 'neoplasia',\n",
       " 217: 'Fainting',\n",
       " 218: 'blood sodium decreased',\n",
       " 219: 'sneezing',\n",
       " 220: 'Feeling unwell',\n",
       " 221: 'hypokalaemia',\n",
       " 222: 'lower GI bleeding',\n",
       " 223: 'migraine',\n",
       " 224: 'intervertebral disc disorder',\n",
       " 225: 'osteoporosis',\n",
       " 226: 'High blood pressure',\n",
       " 227: 'road traffic accident',\n",
       " 228: 'diverticulosis',\n",
       " 229: 'Arthritis rheumatoid',\n",
       " 230: 'Abnormal vision',\n",
       " 231: 'arteriosclerosis',\n",
       " 232: 'bulging',\n",
       " 233: 'Osteoarthritis',\n",
       " 234: 'pleural pain',\n",
       " 235: 'acid reflux',\n",
       " 236: 'fibromyalgia',\n",
       " 237: 'breast pain',\n",
       " 238: 'palpitation',\n",
       " 239: 'herpes zoster',\n",
       " 240: 'Breast cyst',\n",
       " 241: 'myoglobinuria',\n",
       " 242: 'osteomalacia',\n",
       " 243: 'ischias',\n",
       " 244: 'Aching joints',\n",
       " 245: 'arthritis',\n",
       " 246: 'abuse',\n",
       " 247: 'colonic polyp',\n",
       " 248: 'cholecystitis',\n",
       " 249: 'Dyspnea exertional',\n",
       " 250: 'Extrasystoles Ventricular',\n",
       " 251: 'acne rosacea',\n",
       " 252: 'Incontinence',\n",
       " 253: 'gall bladder',\n",
       " 254: 'diabetic acidosis',\n",
       " 255: 'Diabetic Retinopathy',\n",
       " 256: 'blepharitis',\n",
       " 257: 'pyelonephritis',\n",
       " 258: 'aptyalism',\n",
       " 259: 'Blepharospasm',\n",
       " 260: 'aneurysm',\n",
       " 261: 'benign prostatic hyperplasia',\n",
       " 262: 'wrist fracture',\n",
       " 263: 'dislocation',\n",
       " 264: 'Superficial thrombophlebitis',\n",
       " 265: 'Decreased Libido',\n",
       " 266: 'decubitus ulcer',\n",
       " 267: 'kidney failure',\n",
       " 268: 'meningitides',\n",
       " 269: 'ild',\n",
       " 271: 'Atrioventricular block first degree',\n",
       " 272: 'Cough',\n",
       " 273: 'Vestibular disorder',\n",
       " 274: 'panic attack',\n",
       " 275: 'nightmare',\n",
       " 277: 'Breast Lump',\n",
       " 278: 'nocturia',\n",
       " 279: 'spinning sensation',\n",
       " 280: 'Diabetes',\n",
       " 281: 'bronchitis',\n",
       " 282: 'polyp',\n",
       " 283: 'osteomyelitis',\n",
       " 284: 'lung fibrosis',\n",
       " 285: 'enlarged liver',\n",
       " 286: 'laceration',\n",
       " 287: 'coma',\n",
       " 288: 'amyotrophy',\n",
       " 289: 'cachectic',\n",
       " 290: 'renal cyst',\n",
       " 291: 'neuropathy peripheral',\n",
       " 293: 'Bacterial infection',\n",
       " 294: 'angiitis',\n",
       " 295: 'convulsion',\n",
       " 296: 'iron deficiency anaemia',\n",
       " 297: 'Cardiomyopathy',\n",
       " 298: 'bronchial spasm',\n",
       " 299: 'femur fracture',\n",
       " 300: 'fascitis plantar',\n",
       " 301: 'Cardiac decompensation',\n",
       " 302: 'mucositis oral',\n",
       " 303: 'joint swelling',\n",
       " 304: 'hernia',\n",
       " 305: 'glaucoma',\n",
       " 306: 'mitral valve disease NOS',\n",
       " 307: 'edema extremities',\n",
       " 308: 'pulmonary mass',\n",
       " 309: 'breast dysplasia',\n",
       " 310: 'Phlebitis',\n",
       " 311: 'lung neoplasms',\n",
       " 312: 'balance disorder',\n",
       " 313: 'bone marrow failure',\n",
       " 314: 'deglutition disorder',\n",
       " 315: 'adrenal insufficiency',\n",
       " 316: 'sensory disturbance',\n",
       " 317: 'polyneuropathy',\n",
       " 318: 'Nephrolithiasis',\n",
       " 319: 'dysarthria',\n",
       " 320: 'Hypotension Orthostatic',\n",
       " 321: 'Scoliosis',\n",
       " 322: 'autonomic instability',\n",
       " 323: 'rhabdomyolysis',\n",
       " 325: 'bundle branch block left',\n",
       " 326: 'apoplexy',\n",
       " 327: 'malabsorption',\n",
       " 328: 'aphasia',\n",
       " 329: 'Amnesia',\n",
       " 330: 'body temperature increased',\n",
       " 331: 'burning sensation',\n",
       " 332: 'chill',\n",
       " 334: 'Infection Urinary Tract',\n",
       " 335: 'night sweat',\n",
       " 336: 'keratitis sicca',\n",
       " 337: 'dehydration',\n",
       " 338: 'Ataxia',\n",
       " 339: 'transient ischaemic attack',\n",
       " 340: 'cerebral haemorrhage',\n",
       " 341: 'Mod',\n",
       " 342: 'elevated cholesterol',\n",
       " 343: 'heart rate increased',\n",
       " 344: 'eye swelling',\n",
       " 345: 'psychosexual disorder',\n",
       " 346: 'Anxiety',\n",
       " 347: 'facial palsy',\n",
       " 349: 'ureteral obstruction',\n",
       " 350: 'disorder Renal',\n",
       " 351: 'nephrosclerosis',\n",
       " 353: 'psoriatic arthritis',\n",
       " 354: 'post traumatic stress disorder',\n",
       " 355: 'laryngitis',\n",
       " 356: 'angiopathy',\n",
       " 357: 'flu',\n",
       " 358: 'coeliac disease',\n",
       " 359: 'adjustment disorder',\n",
       " 360: 'carcinoid syndrome',\n",
       " 361: 'nausea',\n",
       " 362: 'dysphasia',\n",
       " 363: 'drug withdrawal',\n",
       " 364: 'Pleural Effusion',\n",
       " 365: 'Sinusitis',\n",
       " 366: 'Bilirubinaemia',\n",
       " 367: 'amentia',\n",
       " 368: 'cerebral atrophy',\n",
       " 369: 'cervicalgia',\n",
       " 370: 'hyperlipaemia',\n",
       " 371: 'metabolic encephalopathy',\n",
       " 372: 'Diabetic Nephropathy',\n",
       " 373: 'encephalitis',\n",
       " 374: 'Intervertebral Disc Herniation',\n",
       " 375: 'Atherosclerosis',\n",
       " 376: 'aortic regurgitation',\n",
       " 377: 'Neuralgia',\n",
       " 378: 'hemiplegia',\n",
       " 379: 'pneumonia Klebsiella',\n",
       " 380: 'emphysema',\n",
       " 381: 'AFIB',\n",
       " 382: 'Postherpetic Neuralgia',\n",
       " 383: 'Eosinophil Count Increased',\n",
       " 384: 'Neuritis',\n",
       " 385: 'abnormal ECG',\n",
       " 386: 'Bone Fracture Spontaneous',\n",
       " 387: 'gingivitis',\n",
       " 388: 'soft tissue infection',\n",
       " 389: 'floaters',\n",
       " 390: 'thyroid neoplasia',\n",
       " 391: 'dental abscess',\n",
       " 392: 'dental pain',\n",
       " 393: 'aseptic necrosis bone',\n",
       " 394: 'bruxism',\n",
       " 395: 'retinal bleeding',\n",
       " 396: 'gingival pain',\n",
       " 397: 'bone inflammation',\n",
       " 398: 'candidiasis of mouth',\n",
       " 399: 'tooth disease',\n",
       " 400: 'mouth pain',\n",
       " 401: 'arthropathy',\n",
       " 402: 'pericardial effusion',\n",
       " 403: 'lung neoplasm malignant',\n",
       " 404: 'infection',\n",
       " 405: 'blindness',\n",
       " 406: 'herpes simplex',\n",
       " 407: 'blood in urine',\n",
       " 408: 'Acute myeloblastic leukemia',\n",
       " 409: 'abnormal EEG',\n",
       " 410: 'gastric inflammation',\n",
       " 411: 'Transfusion reaction',\n",
       " 412: 'abortion spontaneous',\n",
       " 413: 'urosepsis',\n",
       " 414: 'eating disorder',\n",
       " 415: 'schizophrenia',\n",
       " 416: 'cardiac murmur',\n",
       " 417: 'cardiac enlargement',\n",
       " 418: 'proteinuria',\n",
       " 419: 'increased platelet count',\n",
       " 420: 'odynophagia',\n",
       " 421: 'multiple sclerosis',\n",
       " 423: 'weight gain',\n",
       " 424: 'Convulsions Grand Mal',\n",
       " 425: 'Enuresis',\n",
       " 426: 'bundle branch block right',\n",
       " 427: 'Personality disorder',\n",
       " 428: 'salivary gland inflammation',\n",
       " 430: 'cervicitis',\n",
       " 431: 'hernia hiatal',\n",
       " 433: 'adenitis',\n",
       " 434: 'ecchymoses',\n",
       " 435: 'tardive dyskinesia',\n",
       " 436: 'blood pressure abnormal',\n",
       " 437: 'muscle strain',\n",
       " 439: 'Ejaculation Premature',\n",
       " 440: 'excessive thirst',\n",
       " 441: 'ear ache',\n",
       " 442: 'elevated erythrocyte sedimentation rate',\n",
       " 443: 'Blood Calcium Increased',\n",
       " 444: 'joint stiffness',\n",
       " 446: 'esophagitis',\n",
       " 447: 'nephrotic syndrome',\n",
       " 448: 'proctalgia',\n",
       " 449: 'bone pain',\n",
       " 450: 'contracture',\n",
       " 451: 'proctitis',\n",
       " 452: 'spondylitis',\n",
       " 453: 'spondylosis',\n",
       " 455: 'hyperparathyroidism',\n",
       " 456: 'sleep apnea',\n",
       " 457: 'Raynauds phenomenon',\n",
       " 458: 'wound dehiscence',\n",
       " 460: 'head injury',\n",
       " 462: 'synovitis',\n",
       " 463: 'hyperhidrosis',\n",
       " 464: 'nystagmus',\n",
       " 465: 'gastrointestinal bleed',\n",
       " 466: 'apraxia',\n",
       " 469: 'Drug addiction',\n",
       " 470: 'prostatitis',\n",
       " 471: 'femoral neck fracture',\n",
       " 472: 'sleep walking',\n",
       " 473: 'monoclonal gammopathy',\n",
       " 474: 'bruise',\n",
       " 475: 'Scar',\n",
       " 476: 'cardiac disease',\n",
       " 477: 'pain pelvic',\n",
       " 478: 'hip fracture',\n",
       " 479: 'difficulty in walking',\n",
       " 480: 'panic disorder',\n",
       " 481: 'atrial septal defect',\n",
       " 482: 'Tendinitis',\n",
       " 483: 'obstructive uropathy',\n",
       " 484: 'Colpitis',\n",
       " 485: 'Rhagades',\n",
       " 486: 'bone spur',\n",
       " 487: 'pancreatitis',\n",
       " 488: 'allergic rhinitis',\n",
       " 489: 'otitis media',\n",
       " 490: 'intestinal functional disorder',\n",
       " 491: 'faecal incontinence',\n",
       " 492: 'carpal tunnel',\n",
       " 493: 'haematochezia',\n",
       " 494: 'ADH inappropriate',\n",
       " 495: 'neuropathy',\n",
       " 496: 'Chest infection',\n",
       " 497: 'radiculopathy',\n",
       " 498: 'dysphonia',\n",
       " 499: 'myelopathy',\n",
       " 500: 'black stools',\n",
       " 501: 'dysuria',\n",
       " 502: 'flashing lights',\n",
       " 503: 'Electrolyte disorder',\n",
       " 504: 'Decreased hearing',\n",
       " 505: 'mitral valve prolapse',\n",
       " 506: 'fatty liver',\n",
       " 507: 'mouth bleeding',\n",
       " 508: 'bipolar disorder',\n",
       " 509: 'embolism',\n",
       " 510: 'duodenitis',\n",
       " 511: 'malignant melanoma',\n",
       " 512: 'ear infection',\n",
       " 513: 'sinus bradycardia',\n",
       " 514: 'atrial ectopic beats',\n",
       " 515: 'external ear infection',\n",
       " 516: 'muscle inflammation',\n",
       " 517: 'cervical vertebral fracture',\n",
       " 518: 'hydronephrosis',\n",
       " 519: 'dental caries',\n",
       " 520: 'elevated prostate specific antigen',\n",
       " 521: 'mouth ulcer',\n",
       " 522: 'gastroduodenal ulcer',\n",
       " 523: 'fistula',\n",
       " 524: 'acute bronchitis',\n",
       " 525: 'gynaecomastia',\n",
       " 526: 'Hypophosphataemia',\n",
       " 527: 'Excess potassium',\n",
       " 528: 'esophageal ulcer',\n",
       " 529: 'renal tubular acidosis',\n",
       " 530: 'phlebothrombosis',\n",
       " 531: 'dysaesthesia',\n",
       " 532: 'goiter',\n",
       " 533: 'metabolic alkalosis',\n",
       " 534: 'Adenopathy',\n",
       " 535: 'tympanic membrane perforation',\n",
       " 536: 'Arrhythmia',\n",
       " 537: 'Hepatic failure',\n",
       " 538: 'hyperalimentation',\n",
       " 539: 'Spinal fracture',\n",
       " 540: 'Parkinson',\n",
       " 541: 'movement disorder',\n",
       " 542: 'macula lutea degeneration',\n",
       " 543: 'musculoskeletal pain',\n",
       " 545: 'Cardiac ischemia',\n",
       " 546: 'paranoia',\n",
       " 548: 'chronic sinusitis',\n",
       " 549: 'ulcer',\n",
       " 550: 'rib fracture',\n",
       " 551: 'rotator cuff syndrome',\n",
       " 553: 'fractured pelvis NOS',\n",
       " 554: 'Infection Upper Respiratory',\n",
       " 555: 'angioedema',\n",
       " 556: 'sick sinus syndrome',\n",
       " 557: 'hyperthyroidism',\n",
       " 558: 'cardiac failure',\n",
       " 559: 'atrioventricular block second degree',\n",
       " 560: 'pharyngeal inflammation',\n",
       " 561: 'adenoma',\n",
       " 562: 'epilepsy',\n",
       " 564: 'angina',\n",
       " 565: 'muscle disorder',\n",
       " 566: 'bundle branch block',\n",
       " 567: 'tachycardia ventricular',\n",
       " 568: 'melanocytic naevus',\n",
       " 569: 'cramp',\n",
       " 570: 'atrial flutter',\n",
       " 571: 'cardiovascular collapse',\n",
       " 572: 'hypernatraemia',\n",
       " 573: 'hive',\n",
       " 575: 'syncope vasovagal',\n",
       " 576: 'streptococcal infection',\n",
       " 577: 'aspergillosis',\n",
       " 578: 'Light sensitivity',\n",
       " 579: 'enteritis',\n",
       " 580: 'bone marrow transplant',\n",
       " 582: 'colitis',\n",
       " 583: 'elevated triglycerides',\n",
       " 584: 'Glucose intolerance',\n",
       " 585: 'Multiple Myeloma',\n",
       " 586: 'duodenal ulcer haemorrhage',\n",
       " 587: 'CMV infection',\n",
       " 588: 'hemiparesis',\n",
       " 589: 'polymyositis',\n",
       " 590: 'Bacteraemia',\n",
       " 591: 'onychomycosis',\n",
       " 592: 'carcinoma of the colon',\n",
       " 594: 'myocarditis',\n",
       " 595: 'cardiac valvulopathy',\n",
       " 596: 'endocarditis',\n",
       " 597: 'Atrioventricular block complete',\n",
       " 598: 'ventricular fibrillation',\n",
       " 599: 'gangrene',\n",
       " 600: 'ischaemia',\n",
       " 601: 'ADVERSE DRUG EFFECT',\n",
       " 602: 'defaecation urgency',\n",
       " 603: 'cirrhosis',\n",
       " 604: 'trigeminal neuralgia',\n",
       " 606: 'polyuria',\n",
       " 607: 'periodontal disease',\n",
       " 608: 'obesity',\n",
       " 609: 'glucosuria',\n",
       " 610: 'basal cell carcinoma',\n",
       " 611: 'pulmonary arrest',\n",
       " 612: 'haematemesis',\n",
       " 613: 'granuloma',\n",
       " 616: 'insulin dependent diabetes mellitus',\n",
       " 617: 'failure to thrive',\n",
       " 618: 'Cardiac tamponade',\n",
       " 619: 'ankle fracture',\n",
       " 620: 'hepatitis toxic',\n",
       " 621: 'endocrine disorder',\n",
       " 622: 'Sinus headache',\n",
       " 624: 'diverticulitis',\n",
       " 625: 'dermatitis exfoliative',\n",
       " 628: 'pancreatitis relapsing',\n",
       " 629: 'Thrombophlebitis',\n",
       " 631: 'shoulder pain',\n",
       " 632: 'Abnormal Laboratory Findings',\n",
       " 634: 'duodenal ulcer perforation',\n",
       " 635: 'joint effusion',\n",
       " 637: 'gouty arthritis',\n",
       " 638: 'Supraventricular tachycardia',\n",
       " 639: 'respiratory alkalosis',\n",
       " 640: 'haemorrhage intracranial',\n",
       " 641: 'major depression',\n",
       " 642: 'excessive sleepiness',\n",
       " 643: 'lymphoma',\n",
       " 644: 'substance abuse',\n",
       " 645: 'animal bite',\n",
       " 646: 'cerebral vascular disorder',\n",
       " 647: 'dry eye',\n",
       " 648: 'Ovarian Cyst',\n",
       " 649: 'micturition urgency',\n",
       " 650: 'Stress incontinence',\n",
       " 651: 'hesitancy',\n",
       " 652: 'bone fracture',\n",
       " 653: 'nodule',\n",
       " 654: 'cyst',\n",
       " 655: 'Abdominal hernia',\n",
       " 656: 'Scleroderma',\n",
       " 657: 'Cor pulmonale',\n",
       " 658: 'excessive salivation',\n",
       " 659: 'right heart failure',\n",
       " 660: 'subarachnoid haemorrhage',\n",
       " 661: 'cancer',\n",
       " 663: 'colitis pseudomembranous',\n",
       " 664: 'heavy menstrual period',\n",
       " 665: 'premature menopause',\n",
       " 666: 'umbilical hernia',\n",
       " 667: 'psoriasis',\n",
       " 668: 'contact dermatitis',\n",
       " 669: 'blepharoptosis',\n",
       " 670: 'Bleeding gums',\n",
       " 671: 'epicondylitis',\n",
       " 672: 'ingrowing nail',\n",
       " 673: 'dermatitis seborrheic',\n",
       " 674: 'candidiasis of vagina',\n",
       " 675: 'Spinal Compression Fracture',\n",
       " 676: 'palsies',\n",
       " 677: 'petit mal',\n",
       " 678: 'gout',\n",
       " 679: 'cerebral ischaemia',\n",
       " 680: 'eyelid diseases',\n",
       " 681: 'gastric ulcer',\n",
       " 682: 'periodontitis',\n",
       " 683: 'Lyell',\n",
       " 684: 'seborrhoeic keratosis',\n",
       " 685: 'actinic keratosis',\n",
       " 686: 'choking',\n",
       " 688: 'aortic stenosis',\n",
       " 689: 'coughing blood',\n",
       " 691: 'Glossitis',\n",
       " 692: 'bacterial pneumonia',\n",
       " 693: 'alcohol consumption',\n",
       " 694: 'squamous cell carcinoma of skin',\n",
       " 695: 'pancreatic cancer',\n",
       " 696: 'Tenosynovitis',\n",
       " 697: 'liver abscess',\n",
       " 698: 'myelodysplasia',\n",
       " 699: 'rectal polyp',\n",
       " 700: 'brain concussion',\n",
       " 701: 'cutaneous ulcer',\n",
       " 703: 'face pain',\n",
       " 704: 'atrioventricular block',\n",
       " 705: 'breast enlargement',\n",
       " 706: 'Eustachian tube disorder',\n",
       " 709: 'acute sinusitis',\n",
       " 710: 'Pelvic Inflammatory Disease',\n",
       " 711: 'Altered Bowel Function',\n",
       " 712: 'antinuclear antibody positive',\n",
       " 713: 'galactorrhea',\n",
       " 714: 'schizoaffective disorder',\n",
       " 715: 'borderline personality disorder',\n",
       " 716: 'injury of neck',\n",
       " 717: 'Pharyngitis Streptococcal',\n",
       " 718: 'Nephrogenic diabetes insipidus',\n",
       " 719: 'hepatic necrosis',\n",
       " 720: 'portal hypertension',\n",
       " 721: 'Diabetic neuropathy',\n",
       " 722: 'brain neoplasm',\n",
       " 723: 'reflux esophagitis',\n",
       " 724: 'Breakthrough bleeding',\n",
       " 726: 'Seroma',\n",
       " 727: 'trigger finger',\n",
       " 728: 'Interstitial nephritis',\n",
       " 729: 'Nail disorder',\n",
       " 730: 'allergic vasculitis',\n",
       " 731: 'lichen planus',\n",
       " 734: 'carotid artery stenosis',\n",
       " 735: 'gastric polyps',\n",
       " 736: 'hyperkeratosis',\n",
       " 739: 'essential hypertension',\n",
       " 740: 'hepatitis B',\n",
       " 742: 'congenital heart disease',\n",
       " 745: 'cholecystectomies',\n",
       " 746: 'Carcinoma of Prostate',\n",
       " 747: 'orthopnea',\n",
       " 748: 'renal mass',\n",
       " 749: 'Breast disorder',\n",
       " 750: 'disseminated lupus erythematosus',\n",
       " 752: 'breast tenderness',\n",
       " 755: 'Disorder Peripheral Vascular',\n",
       " 757: 'Hyperprolactinaemia',\n",
       " 759: 'acute psychosis',\n",
       " 760: 'ketoacidosis',\n",
       " 761: 'sarcoidosis',\n",
       " 762: 'Detached retina',\n",
       " 764: 'Chest wall pain',\n",
       " 767: 'hyperpigmentation',\n",
       " 768: 'costochondritis',\n",
       " 769: 'agoraphobia',\n",
       " 770: 'cheilosis',\n",
       " 771: 'Easy bruisability',\n",
       " 772: 'OPTIC NEURITIS',\n",
       " 773: 'eye injury',\n",
       " 774: 'scotoma',\n",
       " 775: 'eczema',\n",
       " 781: 'erysipelas',\n",
       " 784: 'hepatorenal syndrome',\n",
       " 785: 'Hepatitis C',\n",
       " 786: 'eye infection',\n",
       " 787: 'EBV infection',\n",
       " 788: 'Caesarean Section',\n",
       " 789: 'agranulocytoses',\n",
       " 790: 'petechia',\n",
       " 791: 'hematoma subdural',\n",
       " 793: 'pyoderma gangrenosum',\n",
       " 795: 'cerebral artery embolism',\n",
       " 796: 'hydrocephalus',\n",
       " 798: 'Acne',\n",
       " 799: 'malignant hypertension',\n",
       " 801: 'Post nasal drip',\n",
       " 802: 'labyrinthitis',\n",
       " 803: 'Nodule Skin',\n",
       " 805: 'neck mass',\n",
       " 806: 'tracheitis',\n",
       " 809: 'Vitamin D Deficiency',\n",
       " 810: 'kidney pain',\n",
       " 811: 'hyperuricaemia',\n",
       " 814: 'parotitis',\n",
       " 815: 'choreoathetosis',\n",
       " 816: 'Disorder Retinal',\n",
       " 817: 'pericarditis',\n",
       " 818: 'cerumen impaction',\n",
       " 820: 'paronychia',\n",
       " 822: 'Tooth Impacted',\n",
       " 823: 'thyroid disease',\n",
       " 824: 'aortic aneurysm',\n",
       " 825: 'dermatomyositis',\n",
       " 826: 'hyperaesthesia',\n",
       " 827: 'als',\n",
       " 830: 'bronchiectasis',\n",
       " 833: 'aplasia pure red cell',\n",
       " 835: 'furuncle',\n",
       " 837: 'birth defect',\n",
       " 839: 'autonomic neuropathy',\n",
       " 841: 'polymyalgia rheumatica',\n",
       " 842: 'diaphragmatic hernia',\n",
       " 843: 'lymphedema',\n",
       " 844: 'coronary artery bypass graft',\n",
       " 845: 'pleural fibrosis',\n",
       " 846: 'temporal arteritis',\n",
       " 847: 'polyarthritis',\n",
       " 850: 'fecal occult blood',\n",
       " 851: 'cholecystitis acute',\n",
       " 852: 'inflammatory bowel disease',\n",
       " 853: 'tendon injury',\n",
       " 855: 'hiccough',\n",
       " 856: 'serum sickness',\n",
       " 858: 'Anemia aplastic',\n",
       " 860: 'synovial cyst',\n",
       " 863: 'pneumonia primary atypical',\n",
       " 865: 'Ovarian cancer',\n",
       " 867: 'Gastric Cancer',\n",
       " 869: 'varicose vein',\n",
       " 872: 'aphthous stomatitis',\n",
       " 876: 'Cutaneous mycosis',\n",
       " 878: 'humerus fracture',\n",
       " 879: 'volvulus',\n",
       " 881: 'vaginal discharge',\n",
       " 885: 'left ventricular hypertrophy',\n",
       " 887: 'anosmia',\n",
       " 890: 'haemangioma',\n",
       " 891: 'incisional hernia',\n",
       " 892: 'abnormal mammogram',\n",
       " 893: 'Arthritis bacterial',\n",
       " 894: 'Splenic infarction',\n",
       " 895: 'Hernia Inguinal',\n",
       " 896: 'meningioma',\n",
       " 902: 'hepatic encephalopathy',\n",
       " 903: 'Dyspnoea paroxysmal nocturnal',\n",
       " 904: 'carotid bruit',\n",
       " 905: 'vitamin B 12 deficiency',\n",
       " 907: 'paraplegia',\n",
       " 909: 'Aseptic meningitis',\n",
       " 910: 'Alcoholic intoxication',\n",
       " 911: 'causalgia',\n",
       " 912: 'Irritation Skin',\n",
       " 913: 'Bradyarrhythmia',\n",
       " 914: 'conversion disorder',\n",
       " 915: 'epidural abscess',\n",
       " 916: 'cholecystitis chronic',\n",
       " 918: 'hay fever',\n",
       " 919: 'myasthenia gravis',\n",
       " 920: 'fibroids',\n",
       " 922: 'mediastinal disorder',\n",
       " 923: 'bladder cancer',\n",
       " 926: 'Breast swelling',\n",
       " 930: 'histoplasmosis',\n",
       " 935: 'ganglion',\n",
       " 941: 'Sjogrens syndrome',\n",
       " 943: 'Vitreous haemorrhage',\n",
       " 944: 'venous insufficiency',\n",
       " 947: 'Burns Second Degree',\n",
       " 948: 'tension headache',\n",
       " 949: 'night cramps',\n",
       " 950: 'pigmentation disorder',\n",
       " 951: 'keratitis',\n",
       " 952: 'kidney transplant',\n",
       " 953: 'diabetes insipidus',\n",
       " 954: 'fracture nonunion',\n",
       " 956: 'folliculitis',\n",
       " 958: 'empyema',\n",
       " 960: 'Gastric ulcer perforated',\n",
       " 961: 'platelet disorder',\n",
       " 964: 'allergic alveolitis',\n",
       " 966: 'chronic ulcerative colitis',\n",
       " 967: 'Meningitis Viral',\n",
       " 969: 'esophageal cancer',\n",
       " 975: 'neurogenic bladder',\n",
       " 977: 'uterine bleeding',\n",
       " 978: 'Uterine polyp',\n",
       " 982: 'hyperacusia',\n",
       " 983: 'enterocolitis',\n",
       " 984: 'motor retardation',\n",
       " 985: 'cryptococcosis',\n",
       " 992: 'chronic hepatitis',\n",
       " 993: 'food intolerance',\n",
       " 994: 'cluster headache',\n",
       " 996: 'connective tissue disease',\n",
       " 997: 'macrocytosis',\n",
       " 999: 'Anisocoria',\n",
       " 1002: 'calculus ureteric',\n",
       " 1007: 'Menstrual Disorder',\n",
       " 1008: 'dysphemia',\n",
       " 1010: 'irregular menstrual cycle',\n",
       " 1014: 'pancreatic pseudocyst',\n",
       " 1024: 'testicular swelling',\n",
       " 1031: 'hyperphosphatemia',\n",
       " 1033: 'status epilepticus',\n",
       " 1034: 'sacroiliitis',\n",
       " 1036: 'Tracheobronchitis',\n",
       " 1037: 'pyuria',\n",
       " 1049: 'breast inflammation',\n",
       " 1050: 'xerosis',\n",
       " 1051: 'postmenopausal bleeding',\n",
       " 1062: 'Mumps',\n",
       " 1082: 'tinea pedis',\n",
       " 1085: 'hypogonadism',\n",
       " 1091: 'testes pain',\n",
       " 1095: 'esophageal stenosis',\n",
       " 1107: 'coccydynia',\n",
       " 1112: 'corneal abrasion',\n",
       " 1123: 'Bunion',\n",
       " 1137: 'Viral Pharyngitis',\n",
       " 1138: 'Cystitis Interstitial',\n",
       " 1148: 'asthenopia',\n",
       " 1152: 'endometriosis'}"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping.side_effect_idx_to_name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "drugs = ['CID3883','CID4594','CID4679']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "basepath = 'results/results/out/all-all-536-3.0-0.97'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open('../data/[91.0, 88.0]-[84.0, 84.0]-[28.0, 28.0]-1.0.pkl','rb') as f:\n",
    "#with open('results/out/pkl/all-all-28-1.0-0.99.pkl','rb') as f:\n",
    "#with open('results2/out/pkl/100,34,88-100,34,88-28-1.0-0.5.pkl','rb') as f:\n",
    "#with open('results3/out/pkl/147-all-all-3.0-0.96.pkl','rb') as f:\n",
    "#with open('results3/out/pkl/all-all-508-3.0-0.97.pkl','rb') as f:  # previous\n",
    "#with open('results/results/out/all-all-536-3.0-0.97/exp_info.pkl','rb') as f:\n",
    "with open(os.path.join(basepath, 'exp_info.pkl'),'rb') as f:\n",
    "    data = pickle.load(f)\n",
    "dmap = data.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "idx1 = data['pp_idx'][0]\n",
    "idx2 = data['pp_idx'][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['pp_id'] = [[int(mapping.prot_idx_to_id[idx].replace('GeneID','')) for idx in idx1],\n",
    "                 [int(mapping.prot_idx_to_id[idx].replace('GeneID','')) for idx in idx2]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check the namespace directly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_go = pd.read_csv(os.path.join(basepath, 'name_space.csv'), index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['biological_process', 'cellular_component', 'molecular_function'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go.namespace.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.007001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chloride channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.007055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>steroid hydroxylase activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.007055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>inhibitory extracellular ligand-gated ion channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>benzodiazepine receptor activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>monooxygenase activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>dimethylallyltranstransferase activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>geranyltranstransferase activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>GABA-gated chloride ion channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.010908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>heme binding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.010941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>arachidonic acid epoxygenase activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.014107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>GABA-A receptor activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.027186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>protein homodimerization activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.032488</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                           name  \\\n",
       "32  oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen   \n",
       "33  chloride channel activity                                                                                                                                                                     \n",
       "34  steroid hydroxylase activity                                                                                                                                                                  \n",
       "35  inhibitory extracellular ligand-gated ion channel activity                                                                                                                                    \n",
       "36  benzodiazepine receptor activity                                                                                                                                                              \n",
       "37  monooxygenase activity                                                                                                                                                                        \n",
       "38  dimethylallyltranstransferase activity                                                                                                                                                        \n",
       "39  geranyltranstransferase activity                                                                                                                                                              \n",
       "40  GABA-gated chloride ion channel activity                                                                                                                                                      \n",
       "41  heme binding                                                                                                                                                                                  \n",
       "42  arachidonic acid epoxygenase activity                                                                                                                                                         \n",
       "43  GABA-A receptor activity                                                                                                                                                                      \n",
       "44  protein homodimerization activity                                                                                                                                                             \n",
       "\n",
       "             namespace  p_fdr_bh  \n",
       "32  molecular_function  0.007001  \n",
       "33  molecular_function  0.007055  \n",
       "34  molecular_function  0.007055  \n",
       "35  molecular_function  0.010183  \n",
       "36  molecular_function  0.010183  \n",
       "37  molecular_function  0.010183  \n",
       "38  molecular_function  0.010183  \n",
       "39  molecular_function  0.010183  \n",
       "40  molecular_function  0.010908  \n",
       "41  molecular_function  0.010941  \n",
       "42  molecular_function  0.014107  \n",
       "43  molecular_function  0.027186  \n",
       "44  molecular_function  0.032488  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go[ df_go.namespace.str.contains('molecular')].sort_values('p_fdr_bh')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example 2: Clomipramine --> side effect --> PPI\n",
    "- Drug 1: 147 (Clomipramine)\n",
    "- Drug 2: 24 (all-->one)\n",
    "- Side effect: 376 (aortic regurgitation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze after running below"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "set([mapping.side_effect_idx_to_name[k] for k in data['sd_idx']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['pp_idx', 'pp_weight', 'pd_idx', 'pd_weight', 'drug1', 'durg2', 'side_effect', 'probability', 'pp_id'])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = [mapping.drug_idx_to_id[k] for k in data['durg2']]\n",
    "\n",
    "cids = [int(d.replace('CID','')) for d in d2]\n",
    "drugs = [Compound.from_cid(int(cid)) for cid in set(cids)]\n",
    "\n",
    "drugs\n",
    "\n",
    "drug_ids = [(d.cid, d.synonyms[0], d.iupac_name) for d in drugs]\n",
    "drug_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the side effects to run PoSe-Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "hearts = ['arrhythmia','cardiac','heart','ventric','atri','aort']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('heart attack', 186),\n",
       " ('arteriosclerotic heart disease', 202),\n",
       " ('Extrasystoles Ventricular', 239),\n",
       " ('Atrioventricular block first degree', 259),\n",
       " ('Cardiac decompensation', 287),\n",
       " ('heart rate increased', 327),\n",
       " ('aortic regurgitation', 358),\n",
       " ('cardiac murmur', 398),\n",
       " ('cardiac enlargement', 399),\n",
       " ('cardiac disease', 448),\n",
       " ('atrial septal defect', 453),\n",
       " ('atrial ectopic beats', 486),\n",
       " ('Arrhythmia', 508),\n",
       " ('Cardiac ischemia', 516),\n",
       " ('cardiac failure', 527),\n",
       " ('atrioventricular block second degree', 528),\n",
       " ('tachycardia ventricular', 535),\n",
       " ('atrial flutter', 538),\n",
       " ('cardiac valvulopathy', 560),\n",
       " ('Atrioventricular block complete', 562),\n",
       " ('ventricular fibrillation', 563),\n",
       " ('Cardiac tamponade', 580),\n",
       " ('Supraventricular tachycardia', 594),\n",
       " ('right heart failure', 615),\n",
       " ('aortic stenosis', 642),\n",
       " ('atrioventricular block', 656),\n",
       " ('congenital heart disease', 686),\n",
       " ('aortic aneurysm', 740),\n",
       " ('left ventricular hypertrophy', 773),\n",
       " ('Bradyarrhythmia', 791)]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ses = [(k,v) for k,v in mapping.side_effect_name_to_idx.items() \n",
    " if any([x.lower() in k.lower() for x in hearts])]\n",
    "ses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mapping.side_effect_idx_to_name)\n",
    "name_to_idx = {v:k for k, v in mapping.side_effect_idx_to_name.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'186,202,239,259,287,327,358,398,399,448,453,486,508,516,527,528,535,538,560,562,563,580,594,615,642,656,686,740,773,791'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "','.join([str(x[1]) for x in ses])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Chronic Kidney Disease', 143),\n",
       " ('acute kidney failure', 144),\n",
       " ('kidney failure', 256),\n",
       " ('renal cyst', 277),\n",
       " ('adrenal insufficiency', 301),\n",
       " ('disorder Renal', 333),\n",
       " ('renal tubular acidosis', 501),\n",
       " ('renal mass', 690),\n",
       " ('hepatorenal syndrome', 711),\n",
       " ('kidney pain', 730),\n",
       " ('kidney transplant', 811)]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "renal = ['kidne','renal']\n",
    "ses = [(k,v) for k,v in mapping.side_effect_name_to_idx.items() \n",
    " if any([x.lower() in k.lower() for x in renal])]\n",
    "ses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('muscle spasm', 126),\n",
       " ('muscle paresis', 137),\n",
       " ('aching muscles', 175),\n",
       " ('muscle weakness', 190),\n",
       " ('Cardiomyopathy', 283),\n",
       " ('muscle strain', 416),\n",
       " ('muscle inflammation', 488),\n",
       " ('muscle disorder', 533)]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muscle = ['muscl','myopath']\n",
    "ses = [(k,v) for k,v in mapping.side_effect_name_to_idx.items() \n",
    " if any([x.lower() in k.lower() for x in muscle])]\n",
    "ses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Muscle / Heart / transporter genes\n",
    "https://ascpt.onlinelibrary.wiley.com/doi/abs/10.1038/clpt.2013.234\n",
    "\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pubmed/29463526\n",
    "- potentially strong assocation between arrythmia and SLCO1B1 due to increased risk of prolonged QTc interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GO enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneid2symbol = {v.GeneID: v.Symbol for k,v in GeneID2nt_hum.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "requests.get(http://purl.obolibrary.org/obo/go/go-basic.obo, stream=True)\n",
      "  WROTE: go-basic.obo\n",
      "\n",
      "FTP RETR ftp.ncbi.nlm.nih.gov gene/DATA gene2go.gz -> gene2go.gz\n",
      "  gunzip gene2go.gz\n",
      "go-basic.obo: fmt(1.2) rel(2020-01-01) 47,337 GO Terms\n",
      "HMS:0:00:04.254678 323,107 annotations READ: gene2go \n",
      "1 taxids stored: 9606\n",
      "MF 17,384 annotated human genes\n",
      "CC 18,648 annotated human genes\n",
      "BP 17,541 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 80% 16,711 of 20,913 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 85% 17,755 of 20,913 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 80% 16,699 of 20,913 population items found in association\n"
     ]
    }
   ],
   "source": [
    "\n",
    "obo_fname = download_go_basic_obo()\n",
    "\n",
    "\n",
    "fin_gene2go = download_ncbi_associations()\n",
    "\n",
    "obodag = GODag(\"go-basic.obo\")\n",
    "\n",
    "# Read NCBI's gene2go. Store annotations in a list of namedtuples\n",
    "objanno = Gene2GoReader(fin_gene2go, taxids=[9606])\n",
    "\n",
    "# Get namespace2association where:\n",
    "#    namespace is:\n",
    "#        BP: biological_process               \n",
    "#        MF: molecular_function\n",
    "#        CC: cellular_component\n",
    "#    assocation is a dict:\n",
    "#        key: NCBI GeneID\n",
    "#        value: A set of GO IDs associated with that gene\n",
    "ns2assoc = objanno.get_ns2assc()\n",
    "\n",
    "for nspc, id2gos in ns2assoc.items():\n",
    "    print(\"{NS} {N:,} annotated human genes\".format(NS=nspc, N=len(id2gos)))\n",
    "\n",
    "\n",
    "goeaobj = GOEnrichmentStudyNS(\n",
    "        GeneID2nt_hum.keys(), # List of human protein-acoding genes\n",
    "        ns2assoc, # geneid/GO associations\n",
    "        obodag, # Ontologies\n",
    "        propagate_counts = False,\n",
    "        alpha = 0.05, # default significance cut-off\n",
    "        methods = ['fdr_bh']) # defult multipletest correction method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 94 IDs ...\n",
      " 94%     87 of     93 study items found in association\n",
      " 99%     93 of     94 study items found in population(20913)\n",
      "Calculating 12,189 uncorrected p-values using fisher_scipy_stats\n",
      "  12,189 GO terms are associated with 16,711 of 20,913 population items\n",
      "     693 GO terms are associated with     87 of     94 study items\n",
      "  METHOD fdr_bh:\n",
      "      25 GO terms found significant (< 0.05=alpha) ( 25 enriched +   0 purified): statsmodels fdr_bh\n",
      "      51 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 94 IDs ...\n",
      " 98%     91 of     93 study items found in association\n",
      " 99%     93 of     94 study items found in population(20913)\n",
      "Calculating 1,731 uncorrected p-values using fisher_scipy_stats\n",
      "   1,731 GO terms are associated with 17,755 of 20,913 population items\n",
      "     158 GO terms are associated with     91 of     94 study items\n",
      "  METHOD fdr_bh:\n",
      "       7 GO terms found significant (< 0.05=alpha) (  7 enriched +   0 purified): statsmodels fdr_bh\n",
      "      55 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 94 IDs ...\n",
      " 97%     90 of     93 study items found in association\n",
      " 99%     93 of     94 study items found in population(20913)\n",
      "Calculating 4,164 uncorrected p-values using fisher_scipy_stats\n",
      "   4,164 GO terms are associated with 16,699 of 20,913 population items\n",
      "     264 GO terms are associated with     90 of     94 study items\n",
      "  METHOD fdr_bh:\n",
      "      13 GO terms found significant (< 0.05=alpha) ( 13 enriched +   0 purified): statsmodels fdr_bh\n",
      "      25 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "# 'p_' means \"pvalue\". 'fdr_bh' is the multipletest method we are currently using.\n",
    "# geneids_study = data['pp_id'][0] + data['pp_id'][1] # geneid2symbol.keys()\n",
    "geneids_study = set(data['pp_id'][0] + data['pp_id'][1])\n",
    "goea_results_all = goeaobj.run_study(geneids_study)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "goea_results_sig = [r for r in goea_results_all if r.p_fdr_bh < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "45"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(goea_results_sig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{2222, 2224, 3156, 4047, 4598, 9453, 51478}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0 = goea_results_sig[0]\n",
    "g0.study_items"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cholesterol biosynthetic process</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>steroid metabolic process</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>regulation of cholesterol biosynthetic process</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>0.000024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>xenobiotic metabolic process</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>0.000024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exogenous drug catabolic process</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>0.000094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             name           namespace  \\\n",
       "0  cholesterol biosynthetic process                biological_process   \n",
       "1  steroid metabolic process                       biological_process   \n",
       "2  regulation of cholesterol biosynthetic process  biological_process   \n",
       "3  xenobiotic metabolic process                    biological_process   \n",
       "4  exogenous drug catabolic process                biological_process   \n",
       "\n",
       "   p_fdr_bh  \n",
       "0  0.000005  \n",
       "1  0.000005  \n",
       "2  0.000024  \n",
       "3  0.000024  \n",
       "4  0.000094  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GOTerm('GO:0006695'):\n",
       "  id:GO:0006695\n",
       "  item_id:GO:0006695\n",
       "  name:cholesterol biosynthetic process\n",
       "  namespace:biological_process\n",
       "  _parents: 3 items\n",
       "    GO:1902653\n",
       "    GO:0016126\n",
       "    GO:0008203\n",
       "  parents: 3 items\n",
       "    GO:0008203\tlevel-05\tdepth-06\tcholesterol metabolic process [biological_process]\n",
       "    GO:1902653\tlevel-05\tdepth-06\tsecondary alcohol biosynthetic process [biological_process]\n",
       "    GO:0016126\tlevel-05\tdepth-06\tsterol biosynthetic process [biological_process]\n",
       "  children: 3 items\n",
       "    GO:0033490\tlevel-07\tdepth-08\tcholesterol biosynthetic process via lathosterol [biological_process]\n",
       "    GO:0033488\tlevel-07\tdepth-08\tcholesterol biosynthetic process via 24,25-dihydrolanosterol [biological_process]\n",
       "    GO:0033489\tlevel-07\tdepth-08\tcholesterol biosynthetic process via desmosterol [biological_process]\n",
       "  level:6\n",
       "  depth:7\n",
       "  is_obsolete:False\n",
       "  alt_ids: 0 items"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0 = goea_results_sig[0]\n",
    "g0.goterm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'biological_process'"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0.goterm.__dict__['namespace']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.712216364885412e-06"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0.__dict__['p_fdr_bh']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GO:0006695'"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0 = goea_results_sig[0]\n",
    "g0.goterm.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = ['name','namespace','id']\n",
    "df_go1 = pd.DataFrame([\n",
    "        {k:g.goterm.__dict__.get(k) for k in keys}\n",
    "        for g in goea_results_sig])\n",
    "df_p = pd.DataFrame([{'p_fdr_bh':g.__dict__['p_fdr_bh']} for g in goea_results_sig])\n",
    "df_go = df_go1.merge(df_p, left_index=True, right_index=True)\n",
    "go_genes = pd.DataFrame([{'id':g.goterm.id, 'gene':s,'symbol':geneid2symbol[s]} for g in goea_results_sig for s in g.study_items])\n",
    "df_go = df_go.merge(go_genes, on='id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "      <th>gene</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>namespace</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>biological_process</th>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular_component</th>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>molecular_function</th>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    name   id  p_fdr_bh  gene  symbol\n",
       "namespace                                            \n",
       "biological_process  126   126  126       126   126   \n",
       "cellular_component  102   102  102       102   102   \n",
       "molecular_function  55    55   55        55    55    "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go.groupby('namespace').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "from cobra.io import load_json_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = load_json_model('/home/laurence/models/BiGG_M/json/Recon3D.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 755.003</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10FTHF5GLUtl</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF5GLUtm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF6GLUtl</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF6GLUtm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF7GLUtl</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYOR_u10mi</th>\n",
       "      <td>433.920904</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Htmi</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADH2_u10mi</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYOOm3i</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYOOm2i</th>\n",
       "      <td>566.079096</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10600 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 755.003 at 0x7f6fc62f8b00>"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([  9453,   2222,   4047,   2224,   3156,  51478,   4598,   8608,\n",
       "        54658,   1555,   1557,   1558,   8856,   1565,   2941,  54578,\n",
       "         3326,   9376,   6529,   6532,  10599,   8647,  28234,  10257,\n",
       "        11160,   6554,   5243,   8399,   1071,   9970, 116150,    857,\n",
       "         1536,  55107,    483,   5349,   2555,   2556,   2558,  57107,\n",
       "         9992,   3757,   3269,   3786,   3274,   6623,   7363, 151056,\n",
       "         2515,   7173,   3290,   6534,  10900,  93589,    924,   5578,\n",
       "       134864,  83795,   5590,   1636,    356,   1006,  11120, 440435,\n",
       "         3958,  10993,    818])"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sig = df_go.gene.unique()\n",
    "genes_sig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_gene_ids = [int(g.id.split('_')[0].strip()) for g in model.genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sig_M = [g for g in genes_sig if g in M_gene_ids]\n",
    "len(genes_sig_M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g_rxn = pd.DataFrame([{'gene':int(g.id.split('_')[0].strip()),'rxn':r.reaction,\n",
    "                          'rxn_id':r.id, 'rxn_name':r.name, 'subsystem':r.subsystem}\n",
    "                         for g in model.genes for r in g.reactions])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>rxn</th>\n",
       "      <th>rxn_id</th>\n",
       "      <th>rxn_name</th>\n",
       "      <th>subsystem</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>hdca24g_c &lt;=&gt; hdca24g_r</td>\n",
       "      <td>HDCA24Gtr</td>\n",
       "      <td>Glucuronidated Compound Transport (HDCA-24G)</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8639</td>\n",
       "      <td>h2o_c + o2_c + ptrc_c --&gt; 4abutn_c + h2o2_c + nh4_c</td>\n",
       "      <td>PTRCOX1</td>\n",
       "      <td>Putrescine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8639</td>\n",
       "      <td>dopa_c + h2o_c + o2_c --&gt; 34dhpac_c + h2o2_c + nh4_c</td>\n",
       "      <td>42A12BOOX</td>\n",
       "      <td>4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8639</td>\n",
       "      <td>h2o_c + mhista_c + o2_c --&gt; 3mldz_c + h2o2_c + nh4_c</td>\n",
       "      <td>MHISOR</td>\n",
       "      <td>N-Methylhistamine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8639</td>\n",
       "      <td>13dampp_c + h2o_c + o2_c --&gt; bamppald_c + h2o2_c + nh4_c</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>1,3-Diaminopropane:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Beta-Alanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>8639</td>\n",
       "      <td>h2o_c + mma_c + o2_c --&gt; fald_c + h2o2_c + nh4_c</td>\n",
       "      <td>MAOX</td>\n",
       "      <td>Methylamine:oxygen oxidoreductase (deaminating) (copper-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8639</td>\n",
       "      <td>aact_c + h2o_c + o2_c --&gt; h2o2_c + mthgxl_c + nh4_c</td>\n",
       "      <td>AACTOOR</td>\n",
       "      <td>Aminoacetone:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Glycine, serine, alanine, and threonine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + peamn_c --&gt; h2o2_c + nh4_c + pacald_c</td>\n",
       "      <td>PEAMNO</td>\n",
       "      <td>Phenethylamine oxidase</td>\n",
       "      <td>Phenylalanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>314</td>\n",
       "      <td>dopa_c + h2o_c + o2_c --&gt; 34dhpac_c + h2o2_c + nh4_c</td>\n",
       "      <td>42A12BOOX</td>\n",
       "      <td>4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + mma_c + o2_c --&gt; fald_c + h2o2_c + nh4_c</td>\n",
       "      <td>MAOX</td>\n",
       "      <td>Methylamine:oxygen oxidoreductase (deaminating) (copper-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + trypta_c --&gt; h2o2_c + id3acald_c + nh4_c</td>\n",
       "      <td>TRYPTAOX</td>\n",
       "      <td>Tryptamine:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tryptophan metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + hista_c + o2_c --&gt; h2o2_c + im4act_c + nh4_c</td>\n",
       "      <td>HISTASE</td>\n",
       "      <td>HISTASE</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + ptrc_c --&gt; 4abutn_c + h2o2_c + nh4_c</td>\n",
       "      <td>PTRCOX1</td>\n",
       "      <td>Putrescine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + mhista_c + o2_c --&gt; 3mldz_c + h2o2_c + nh4_c</td>\n",
       "      <td>MHISOR</td>\n",
       "      <td>N-Methylhistamine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>314</td>\n",
       "      <td>13dampp_c + h2o_c + o2_c --&gt; bamppald_c + h2o2_c + nh4_c</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>1,3-Diaminopropane:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Beta-Alanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + peamn_c --&gt; h2o2_c + nh4_c + pacald_c</td>\n",
       "      <td>PEAMNO</td>\n",
       "      <td>Phenethylamine oxidase</td>\n",
       "      <td>Phenylalanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>314</td>\n",
       "      <td>dopa_c + h2o_c + o2_c --&gt; 34dhpac_c + h2o2_c + nh4_c</td>\n",
       "      <td>42A12BOOX</td>\n",
       "      <td>4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + mma_c + o2_c --&gt; fald_c + h2o2_c + nh4_c</td>\n",
       "      <td>MAOX</td>\n",
       "      <td>Methylamine:oxygen oxidoreductase (deaminating) (copper-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + trypta_c --&gt; h2o2_c + id3acald_c + nh4_c</td>\n",
       "      <td>TRYPTAOX</td>\n",
       "      <td>Tryptamine:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tryptophan metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + hista_c + o2_c --&gt; h2o2_c + im4act_c + nh4_c</td>\n",
       "      <td>HISTASE</td>\n",
       "      <td>HISTASE</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + ptrc_c --&gt; 4abutn_c + h2o2_c + nh4_c</td>\n",
       "      <td>PTRCOX1</td>\n",
       "      <td>Putrescine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + mhista_c + o2_c --&gt; 3mldz_c + h2o2_c + nh4_c</td>\n",
       "      <td>MHISOR</td>\n",
       "      <td>N-Methylhistamine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>314</td>\n",
       "      <td>13dampp_c + h2o_c + o2_c --&gt; bamppald_c + h2o2_c + nh4_c</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>1,3-Diaminopropane:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Beta-Alanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>26</td>\n",
       "      <td>h2o_c + o2_c + peamn_c --&gt; h2o2_c + nh4_c + pacald_c</td>\n",
       "      <td>PEAMNO</td>\n",
       "      <td>Phenethylamine oxidase</td>\n",
       "      <td>Phenylalanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>26</td>\n",
       "      <td>dopa_c + h2o_c + o2_c --&gt; 34dhpac_c + h2o2_c + nh4_c</td>\n",
       "      <td>42A12BOOX</td>\n",
       "      <td>4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>h2o_e + o2_e + spmd_e --&gt; CE1939_e + h2o2_e + nh4_e</td>\n",
       "      <td>RE0828E</td>\n",
       "      <td>RE0828</td>\n",
       "      <td>Urea cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>h2o_c + mma_c + o2_c --&gt; fald_c + h2o2_c + nh4_c</td>\n",
       "      <td>MAOX</td>\n",
       "      <td>Methylamine:oxygen oxidoreductase (deaminating) (copper-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>26</td>\n",
       "      <td>h2o_e + o2_e + ptrc_e --&gt; 4abutn_e + h2o2_e + nh4_e</td>\n",
       "      <td>r0281</td>\n",
       "      <td>Putrescine:oxygen oxidoreductase (deaminating) Urea cycle and metabolism of amino groups EC:1.4.3.6</td>\n",
       "      <td>Methionine and cysteine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>26</td>\n",
       "      <td>CE1940_e + h2o_e + o2_e --&gt; CE1943_e + h2o2_e + nh4_e</td>\n",
       "      <td>RE0690E</td>\n",
       "      <td>RE0690</td>\n",
       "      <td>Urea cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>26</td>\n",
       "      <td>h2o_c + o2_c + trypta_c --&gt; h2o2_c + id3acald_c + nh4_c</td>\n",
       "      <td>TRYPTAOX</td>\n",
       "      <td>Tryptamine:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tryptophan metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14457</th>\n",
       "      <td>1351</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14458</th>\n",
       "      <td>1351</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14459</th>\n",
       "      <td>84701</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14460</th>\n",
       "      <td>1347</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14461</th>\n",
       "      <td>1347</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14462</th>\n",
       "      <td>1340</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14463</th>\n",
       "      <td>1340</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14464</th>\n",
       "      <td>4512</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14465</th>\n",
       "      <td>1349</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14466</th>\n",
       "      <td>1349</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14467</th>\n",
       "      <td>1339</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14468</th>\n",
       "      <td>1339</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14469</th>\n",
       "      <td>125965</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14470</th>\n",
       "      <td>1337</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14471</th>\n",
       "      <td>1337</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14472</th>\n",
       "      <td>9377</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14473</th>\n",
       "      <td>9377</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14474</th>\n",
       "      <td>4513</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14475</th>\n",
       "      <td>1329</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14476</th>\n",
       "      <td>1329</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14477</th>\n",
       "      <td>1346</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14478</th>\n",
       "      <td>170712</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14479</th>\n",
       "      <td>170712</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14480</th>\n",
       "      <td>9167</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14481</th>\n",
       "      <td>1345</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14482</th>\n",
       "      <td>1345</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14483</th>\n",
       "      <td>1350</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14484</th>\n",
       "      <td>1350</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14485</th>\n",
       "      <td>1327</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14486</th>\n",
       "      <td>1327</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14487 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  \\\n",
       "0      0      \n",
       "1      8639   \n",
       "2      8639   \n",
       "3      8639   \n",
       "4      8639   \n",
       "...     ...   \n",
       "14482  1345   \n",
       "14483  1350   \n",
       "14484  1350   \n",
       "14485  1327   \n",
       "14486  1327   \n",
       "\n",
       "                                                                                       rxn  \\\n",
       "0      hdca24g_c <=> hdca24g_r                                                               \n",
       "1      h2o_c + o2_c + ptrc_c --> 4abutn_c + h2o2_c + nh4_c                                   \n",
       "2      dopa_c + h2o_c + o2_c --> 34dhpac_c + h2o2_c + nh4_c                                  \n",
       "3      h2o_c + mhista_c + o2_c --> 3mldz_c + h2o2_c + nh4_c                                  \n",
       "4      13dampp_c + h2o_c + o2_c --> bamppald_c + h2o2_c + nh4_c                              \n",
       "...                                                         ...                              \n",
       "14482  4.0 focytC_m + 8.0 h_m + o2_m --> 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i                  \n",
       "14483  4.0 focytC_m + 7.92 h_m + o2_m --> 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m   \n",
       "14484  4.0 focytC_m + 8.0 h_m + o2_m --> 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i                  \n",
       "14485  4.0 focytC_m + 7.92 h_m + o2_m --> 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m   \n",
       "14486  4.0 focytC_m + 8.0 h_m + o2_m --> 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i                  \n",
       "\n",
       "          rxn_id  \\\n",
       "0      HDCA24Gtr   \n",
       "1      PTRCOX1     \n",
       "2      42A12BOOX   \n",
       "3      MHISOR      \n",
       "4      13DAMPPOX   \n",
       "...          ...   \n",
       "14482  CYOOm2i     \n",
       "14483  CYOOm3i     \n",
       "14484  CYOOm2i     \n",
       "14485  CYOOm3i     \n",
       "14486  CYOOm2i     \n",
       "\n",
       "                                                                                     rxn_name  \\\n",
       "0      Glucuronidated Compound Transport (HDCA-24G)                                             \n",
       "1      Putrescine:oxygen oxidoreductase (deaminating)                                           \n",
       "2      4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)   \n",
       "3      N-Methylhistamine:oxygen oxidoreductase (deaminating)                                    \n",
       "4      1,3-Diaminopropane:oxygen oxidoreductase (deaminating)                                   \n",
       "...                                                       ...                                   \n",
       "14482  CYOOm2i                                                                                  \n",
       "14483  CYOOm3i                                                                                  \n",
       "14484  CYOOm2i                                                                                  \n",
       "14485  CYOOm3i                                                                                  \n",
       "14486  CYOOm2i                                                                                  \n",
       "\n",
       "                             subsystem  \n",
       "0      Transport, extracellular         \n",
       "1      Arginine and proline metabolism  \n",
       "2      Tyrosine metabolism              \n",
       "3      Histidine metabolism             \n",
       "4      Beta-Alanine metabolism          \n",
       "...                        ...          \n",
       "14482  Oxidative phosphorylation        \n",
       "14483  Oxidative phosphorylation        \n",
       "14484  Oxidative phosphorylation        \n",
       "14485  Oxidative phosphorylation        \n",
       "14486  Oxidative phosphorylation        \n",
       "\n",
       "[14487 rows x 5 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g_rxn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "sol = model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sol.fluxes['35DSMVhep']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/dist-packages/pandas/core/series.py:1148 \u001b[1;31mFutureWarning\u001b[0m: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "NaN          NaN         \n",
       "RDH1          0.000000   \n",
       "HMGCOARx      50.724638  \n",
       "P4502C94      0.000000   \n",
       "RE2235R       0.000000   \n",
       "                ...      \n",
       "RE2147C       1000.000000\n",
       "SRTNt6_2_r   -500.000000 \n",
       "6CSMVhep      0.000000   \n",
       "HMR_0942      0.000000   \n",
       "UGT1A2r       0.000000   \n",
       "Name: fluxes, Length: 80, dtype: float64"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sol.fluxes[set(df_go_mol.rxn_id.unique())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_go.loc[:,'in_M'] = df_go.gene.isin(genes_sig_M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_go_rxn = pd.merge(df_go, df_g_rxn, on='gene', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "      <th>symbol</th>\n",
       "      <th>rxn</th>\n",
       "      <th>rxn_id</th>\n",
       "      <th>rxn_name</th>\n",
       "      <th>subsystem</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2772</th>\n",
       "      <td>oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen</td>\n",
       "      <td>0.007001</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>h_r + nadph_r + o2_r + smv_r --&gt; 35dsmv_r + h2o_r + nadp_r</td>\n",
       "      <td>35DSMVhep</td>\n",
       "      <td>Formation of 3,5-dihydrodoil simvastatin</td>\n",
       "      <td>Drug metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2773</th>\n",
       "      <td>oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen</td>\n",
       "      <td>0.007001</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>6hsmv_r --&gt; 3hsmv_r</td>\n",
       "      <td>3HSMVhep</td>\n",
       "      <td>Acid catalyzed rearrangement of 6-beta-hydroxy-simvastatin to 3-hydroxy simvastatin</td>\n",
       "      <td>Drug metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2774</th>\n",
       "      <td>oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen</td>\n",
       "      <td>0.007001</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>h_r + nadph_r + o2_r + smv_r --&gt; 6msmv_r + 2.0 h2o_r + nadp_r</td>\n",
       "      <td>6MSMVhep</td>\n",
       "      <td>Oxidation of simvastatin to 6-exomethylene lactone form</td>\n",
       "      <td>Drug metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2775</th>\n",
       "      <td>oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen</td>\n",
       "      <td>0.007001</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>glz_r + h_r + nadph_r + o2_r --&gt; 6bhglz_r + h2o_r + nadp_r</td>\n",
       "      <td>6BHGLZhr</td>\n",
       "      <td>Oxidation of gliclazide to 6-beta-OH-gliclazide in hepatocytes</td>\n",
       "      <td>Drug metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2776</th>\n",
       "      <td>oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen</td>\n",
       "      <td>0.007001</td>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>estrone_r + h_r + nadph_r + o2_r --&gt; C05300_r + h2o_r + nadp_r</td>\n",
       "      <td>RE2235R</td>\n",
       "      <td>Unspecific Monooxygenase</td>\n",
       "      <td>Androgen and estrogen synthesis and metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                             name  \\\n",
       "2772  oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen   \n",
       "2773  oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen   \n",
       "2774  oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen   \n",
       "2775  oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen   \n",
       "2776  oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen, reduced flavin or flavoprotein as one donor, and incorporation of one atom of oxygen   \n",
       "\n",
       "      p_fdr_bh  symbol  \\\n",
       "2772  0.007001  CYP2D6   \n",
       "2773  0.007001  CYP2D6   \n",
       "2774  0.007001  CYP2D6   \n",
       "2775  0.007001  CYP2D6   \n",
       "2776  0.007001  CYP2D6   \n",
       "\n",
       "                                                                 rxn  \\\n",
       "2772  h_r + nadph_r + o2_r + smv_r --> 35dsmv_r + h2o_r + nadp_r       \n",
       "2773  6hsmv_r --> 3hsmv_r                                              \n",
       "2774  h_r + nadph_r + o2_r + smv_r --> 6msmv_r + 2.0 h2o_r + nadp_r    \n",
       "2775  glz_r + h_r + nadph_r + o2_r --> 6bhglz_r + h2o_r + nadp_r       \n",
       "2776  estrone_r + h_r + nadph_r + o2_r --> C05300_r + h2o_r + nadp_r   \n",
       "\n",
       "         rxn_id  \\\n",
       "2772  35DSMVhep   \n",
       "2773  3HSMVhep    \n",
       "2774  6MSMVhep    \n",
       "2775  6BHGLZhr    \n",
       "2776  RE2235R     \n",
       "\n",
       "                                                                                 rxn_name  \\\n",
       "2772  Formation of 3,5-dihydrodoil simvastatin                                              \n",
       "2773  Acid catalyzed rearrangement of 6-beta-hydroxy-simvastatin to 3-hydroxy simvastatin   \n",
       "2774  Oxidation of simvastatin to 6-exomethylene lactone form                               \n",
       "2775  Oxidation of gliclazide to 6-beta-OH-gliclazide in hepatocytes                        \n",
       "2776  Unspecific Monooxygenase                                                              \n",
       "\n",
       "                                           subsystem  \n",
       "2772  Drug metabolism                                 \n",
       "2773  Drug metabolism                                 \n",
       "2774  Drug metabolism                                 \n",
       "2775  Drug metabolism                                 \n",
       "2776  Androgen and estrogen synthesis and metabolism  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_mol = df_go_rxn[df_go_rxn.namespace.str.contains('molecular')].sort_values(['p_fdr_bh','namespace','name'])\n",
    "cols = ['name','p_fdr_bh','symbol','rxn','rxn_id','rxn_name','subsystem']\n",
    "df_go_mol[df_go_mol.in_M][cols].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p_fdr_bh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subsystem</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Fatty acid oxidation</th>\n",
       "      <td>0.008619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Androgen and estrogen synthesis and metabolism</th>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tryptophan metabolism</th>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steroid metabolism</th>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N-glycan metabolism</th>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Limonene and pinene degradation</th>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linoleate metabolism</th>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cytochrome metabolism</th>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Arachidonic acid metabolism</th>\n",
       "      <td>0.010183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Drug metabolism</th>\n",
       "      <td>0.010562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tyrosine metabolism</th>\n",
       "      <td>0.010941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Purine catabolism</th>\n",
       "      <td>0.010941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cholesterol metabolism</th>\n",
       "      <td>0.021336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Methionine and cysteine metabolism</th>\n",
       "      <td>0.032488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Miscellaneous</th>\n",
       "      <td>0.032488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Transport, extracellular</th>\n",
       "      <td>0.032488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Glycine, serine, alanine, and threonine metabolism</th>\n",
       "      <td>0.032488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vitamin A metabolism</th>\n",
       "      <td>0.032488</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    p_fdr_bh\n",
       "subsystem                                                   \n",
       "Fatty acid oxidation                                0.008619\n",
       "Androgen and estrogen synthesis and metabolism      0.010183\n",
       "Tryptophan metabolism                               0.010183\n",
       "Steroid metabolism                                  0.010183\n",
       "N-glycan metabolism                                 0.010183\n",
       "Limonene and pinene degradation                     0.010183\n",
       "Linoleate metabolism                                0.010183\n",
       "Cytochrome metabolism                               0.010183\n",
       "Arachidonic acid metabolism                         0.010183\n",
       "Drug metabolism                                     0.010562\n",
       "Tyrosine metabolism                                 0.010941\n",
       "Purine catabolism                                   0.010941\n",
       "Cholesterol metabolism                              0.021336\n",
       "Methionine and cysteine metabolism                  0.032488\n",
       "Miscellaneous                                       0.032488\n",
       "Transport, extracellular                            0.032488\n",
       "Glycine, serine, alanine, and threonine metabolism  0.032488\n",
       "Vitamin A metabolism                                0.032488"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols2 = ['p_fdr_bh','subsystem']\n",
    "df_go_mol[df_go_mol.in_M][cols2].drop_duplicates().groupby('subsystem').median().sort_values('p_fdr_bh')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Literature validation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-  \"alterations in FA metabolism could play a pivotal role in production of arrhythmias\":\n",
    "https://www.fasebj.org/doi/abs/10.1096/fasebj.31.1_supplement.782.14\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>id</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "      <th>gene</th>\n",
       "      <th>symbol</th>\n",
       "      <th>in_M</th>\n",
       "      <th>rxn</th>\n",
       "      <th>rxn_id</th>\n",
       "      <th>rxn_name</th>\n",
       "      <th>subsystem</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3233</th>\n",
       "      <td>protein homodimerization activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0042803</td>\n",
       "      <td>0.032488</td>\n",
       "      <td>3156</td>\n",
       "      <td>HMGCR</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0 h_x + hmgcoa_x + 2.0 nadph_x --&gt; coa_x + mev__R_x + 2.0 nadp_x</td>\n",
       "      <td>HMGCOARx</td>\n",
       "      <td>Hydroxymethylglutaryl CoA reductase (ir)</td>\n",
       "      <td>Cholesterol metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   name           namespace          id  \\\n",
       "3233  protein homodimerization activity  molecular_function  GO:0042803   \n",
       "\n",
       "      p_fdr_bh  gene symbol  in_M  \\\n",
       "3233  0.032488  3156  HMGCR  True   \n",
       "\n",
       "                                                                     rxn  \\\n",
       "3233  2.0 h_x + hmgcoa_x + 2.0 nadph_x --> coa_x + mev__R_x + 2.0 nadp_x   \n",
       "\n",
       "        rxn_id                                  rxn_name  \\\n",
       "3233  HMGCOARx  Hydroxymethylglutaryl CoA reductase (ir)   \n",
       "\n",
       "                   subsystem  \n",
       "3233  Cholesterol metabolism  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_mol[ df_go_mol.rxn_id.isin(['HMGCOARx'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10600"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.reactions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5835"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.metabolites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Drug metabolism',\n",
       "       'Androgen and estrogen synthesis and metabolism',\n",
       "       'Arachidonic acid metabolism', 'Steroid metabolism',\n",
       "       'Fatty acid oxidation', 'Linoleate metabolism',\n",
       "       'Cytochrome metabolism', 'Tryptophan metabolism',\n",
       "       'Limonene and pinene degradation', nan, 'N-glycan metabolism',\n",
       "       'Cholesterol metabolism', 'Tyrosine metabolism',\n",
       "       'Purine catabolism', 'Vitamin A metabolism', 'Miscellaneous',\n",
       "       'Transport, extracellular',\n",
       "       'Glycine, serine, alanine, and threonine metabolism',\n",
       "       'Methionine and cysteine metabolism'], dtype=object)"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_mol.subsystem.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>id</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "      <th>gene</th>\n",
       "      <th>symbol</th>\n",
       "      <th>in_M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>chloride channel complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0034707</td>\n",
       "      <td>0.016290</td>\n",
       "      <td>55107</td>\n",
       "      <td>ANO1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>chloride channel complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0034707</td>\n",
       "      <td>0.016290</td>\n",
       "      <td>2556</td>\n",
       "      <td>GABRA3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224</th>\n",
       "      <td>chloride channel complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0034707</td>\n",
       "      <td>0.016290</td>\n",
       "      <td>2558</td>\n",
       "      <td>GABRA5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>chloride channel complex</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>GO:0034707</td>\n",
       "      <td>0.016290</td>\n",
       "      <td>2555</td>\n",
       "      <td>GABRA2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>chloride channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0005254</td>\n",
       "      <td>0.007055</td>\n",
       "      <td>55107</td>\n",
       "      <td>ANO1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>chloride channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0005254</td>\n",
       "      <td>0.007055</td>\n",
       "      <td>5349</td>\n",
       "      <td>FXYD3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>chloride channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0005254</td>\n",
       "      <td>0.007055</td>\n",
       "      <td>2555</td>\n",
       "      <td>GABRA2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>chloride channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0005254</td>\n",
       "      <td>0.007055</td>\n",
       "      <td>2556</td>\n",
       "      <td>GABRA3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>chloride channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0005254</td>\n",
       "      <td>0.007055</td>\n",
       "      <td>2558</td>\n",
       "      <td>GABRA5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244</th>\n",
       "      <td>inhibitory extracellular ligand-gated ion channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0005237</td>\n",
       "      <td>0.010183</td>\n",
       "      <td>2555</td>\n",
       "      <td>GABRA2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>inhibitory extracellular ligand-gated ion channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0005237</td>\n",
       "      <td>0.010183</td>\n",
       "      <td>2556</td>\n",
       "      <td>GABRA3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>246</th>\n",
       "      <td>inhibitory extracellular ligand-gated ion channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0005237</td>\n",
       "      <td>0.010183</td>\n",
       "      <td>2558</td>\n",
       "      <td>GABRA5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>GABA-gated chloride ion channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0022851</td>\n",
       "      <td>0.010908</td>\n",
       "      <td>2555</td>\n",
       "      <td>GABRA2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>GABA-gated chloride ion channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0022851</td>\n",
       "      <td>0.010908</td>\n",
       "      <td>2556</td>\n",
       "      <td>GABRA3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>257</th>\n",
       "      <td>GABA-gated chloride ion channel activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>GO:0022851</td>\n",
       "      <td>0.010908</td>\n",
       "      <td>2558</td>\n",
       "      <td>GABRA5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           name  \\\n",
       "222  chloride channel complex                                     \n",
       "223  chloride channel complex                                     \n",
       "224  chloride channel complex                                     \n",
       "225  chloride channel complex                                     \n",
       "232  chloride channel activity                                    \n",
       "233  chloride channel activity                                    \n",
       "234  chloride channel activity                                    \n",
       "235  chloride channel activity                                    \n",
       "236  chloride channel activity                                    \n",
       "244  inhibitory extracellular ligand-gated ion channel activity   \n",
       "245  inhibitory extracellular ligand-gated ion channel activity   \n",
       "246  inhibitory extracellular ligand-gated ion channel activity   \n",
       "255  GABA-gated chloride ion channel activity                     \n",
       "256  GABA-gated chloride ion channel activity                     \n",
       "257  GABA-gated chloride ion channel activity                     \n",
       "\n",
       "              namespace          id  p_fdr_bh   gene  symbol   in_M  \n",
       "222  cellular_component  GO:0034707  0.016290  55107  ANO1    False  \n",
       "223  cellular_component  GO:0034707  0.016290  2556   GABRA3  False  \n",
       "224  cellular_component  GO:0034707  0.016290  2558   GABRA5  False  \n",
       "225  cellular_component  GO:0034707  0.016290  2555   GABRA2  False  \n",
       "232  molecular_function  GO:0005254  0.007055  55107  ANO1    False  \n",
       "233  molecular_function  GO:0005254  0.007055  5349   FXYD3   False  \n",
       "234  molecular_function  GO:0005254  0.007055  2555   GABRA2  False  \n",
       "235  molecular_function  GO:0005254  0.007055  2556   GABRA3  False  \n",
       "236  molecular_function  GO:0005254  0.007055  2558   GABRA5  False  \n",
       "244  molecular_function  GO:0005237  0.010183  2555   GABRA2  False  \n",
       "245  molecular_function  GO:0005237  0.010183  2556   GABRA3  False  \n",
       "246  molecular_function  GO:0005237  0.010183  2558   GABRA5  False  \n",
       "255  molecular_function  GO:0022851  0.010908  2555   GABRA2  False  \n",
       "256  molecular_function  GO:0022851  0.010908  2556   GABRA3  False  \n",
       "257  molecular_function  GO:0022851  0.010908  2558   GABRA5  False  "
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go[df_go.name.str.contains('channel')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "153"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping.side_effect_name_to_idx['Chronic Kidney Disease']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Last step: show using simulation the mechanism by which FA oxidation (and others) cause the Arrhythmia side effect\n",
    "\n",
    "Steps:\n",
    "1. make heart-specific model\n",
    "1. simulate healthy heart metabolism (linked to electrical conduction)\n",
    "1. simulate arrhythmia\n",
    "    1. by perturbing the PoSe-Path genes. Hypothesis: perturbing these genes causes arrhythmia, or that arrhythmia involves change in the reactions encoded by these genes\n",
    "    1. by using known arrhythmia simulation parameters, and check consistency of known arrhythmia mechanisms with the genes found by PoSe-Path [X: DON'T already know arrhythmia parameters]\n",
    "1. Go with A.\n",
    "    - Validate using another metabolic marker of arrhythmia--show that it changes in the correct direction\n",
    "    - Validate using omics (RNA-Seq, proteomics, metabolomics) data from arrhythmia patients\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 755.003</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10FTHF5GLUtl</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF5GLUtm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF6GLUtl</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF6GLUtm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF7GLUtl</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYOR_u10mi</th>\n",
       "      <td>433.920904</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Htmi</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADH2_u10mi</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYOOm3i</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYOOm2i</th>\n",
       "      <td>566.079096</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10600 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 755.003 at 0x7f6fc25d5c88>"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sol"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneid2info = {v.GeneID: (v.Symbol,v.description) for k,v in GeneID2nt_hum.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('CYBB', 'cytochrome b-245, beta polypeptide'),\n",
       " ('TPO', 'thyroid peroxidase'),\n",
       " ('PLB1', 'phospholipase B1'),\n",
       " ('ABCC4', 'ATP binding cassette subfamily C member 4'),\n",
       " ('CYP2B6', 'cytochrome P450 family 2 subfamily B member 6'),\n",
       " ('CYP2C19', 'cytochrome P450 family 2 subfamily C member 19'),\n",
       " ('CYP2C8', 'cytochrome P450 family 2 subfamily C member 8'),\n",
       " ('CYP2D6', 'cytochrome P450 family 2 subfamily D member 6'),\n",
       " ('PDHA2', 'pyruvate dehydrogenase (lipoamide) alpha 2'),\n",
       " ('CETP', 'cholesteryl ester transfer protein, plasma'),\n",
       " ('RUSC1', 'RUN and SH3 domain containing 1'),\n",
       " ('ZSWIM2', 'zinc finger SWIM-type containing 2'),\n",
       " ('SLCO1B3', 'solute carrier organic anion transporter family member 1B3'),\n",
       " ('BCAT2', 'branched chain amino-acid transaminase 2, mitochondrial'),\n",
       " ('TSHB', 'thyroid stimulating hormone beta'),\n",
       " ('HMGCR', '3-hydroxy-3-methylglutaryl-CoA reductase'),\n",
       " ('ACE', 'angiotensin I converting enzyme'),\n",
       " ('TMEM52B', 'transmembrane protein 52B'),\n",
       " ('GPR179', 'G protein-coupled receptor 179'),\n",
       " ('ABCB1', 'ATP binding cassette subfamily B member 1'),\n",
       " ('MED24', 'mediator complex subunit 24'),\n",
       " ('RBP7', 'retinol binding protein 7'),\n",
       " ('RUNDC3A', 'RUN domain containing 3A'),\n",
       " ('UBA7', 'ubiquitin like modifier activating enzyme 7'),\n",
       " ('NR1I2', 'nuclear receptor subfamily 1 group I member 2'),\n",
       " ('SLC22A8', 'solute carrier family 22 member 8'),\n",
       " ('KCNH2', 'potassium voltage-gated channel subfamily H member 2'),\n",
       " ('FDFT1', 'farnesyl-diphosphate farnesyltransferase 1'),\n",
       " ('FDPS', 'farnesyl diphosphate synthase'),\n",
       " ('SLC35A2',\n",
       "  'solute carrier family 35 (UDP-galactose transporter), member A2'),\n",
       " ('UGT2B4', 'UDP glucuronosyltransferase 2 family, polypeptide B4'),\n",
       " ('HRH1', 'histamine receptor H1'),\n",
       " ('FAM189A2', 'family with sequence similarity 189 member A2'),\n",
       " ('HRH2', 'histamine receptor H2'),\n",
       " ('KCNQ3', 'potassium voltage-gated channel subfamily Q member 3'),\n",
       " ('PLA2G10', 'phospholipase A2 group X'),\n",
       " ('TAAR1', 'trace amine associated receptor 1'),\n",
       " ('HSD11B1', 'hydroxysteroid (11-beta) dehydrogenase 1'),\n",
       " ('FXYD3', 'FXYD domain containing ion transport regulator 3'),\n",
       " ('CCDC109B', 'coiled-coil domain containing 109B'),\n",
       " ('GGPS1', 'geranylgeranyl diphosphate synthase 1'),\n",
       " ('DNASE1L1', 'deoxyribonuclease I-like 1'),\n",
       " ('SDS', 'serine dehydratase'),\n",
       " ('NR1I3', 'nuclear receptor subfamily 1 group I member 3'),\n",
       " ('HSP90AB1', 'heat shock protein 90kDa alpha family class B member 1'),\n",
       " ('DMTF1', 'cyclin D binding myb like transcription factor 1'),\n",
       " ('KCNE2', 'potassium voltage-gated channel subfamily E regulatory subunit 2'),\n",
       " ('PDSS2', 'prenyl (decaprenyl) diphosphate synthase, subunit 2'),\n",
       " ('HSD17B7', 'hydroxysteroid (17-beta) dehydrogenase 7'),\n",
       " ('OXLD1', 'oxidoreductase-like domain containing 1'),\n",
       " ('C11orf54', 'chromosome 11 open reading frame 54'),\n",
       " ('RBP5', 'retinol binding protein 5'),\n",
       " ('CAMK2G', 'calcium/calmodulin-dependent protein kinase II gamma'),\n",
       " ('UGT1A6', 'UDP glucuronosyltransferase 1 family, polypeptide A6'),\n",
       " ('ANO1', 'anoctamin 1, calcium activated chloride channel'),\n",
       " ('KCNK16', 'potassium two pore domain channel subfamily K member 16'),\n",
       " ('CACHD1', 'cache domain containing 1'),\n",
       " ('CAV1', 'caveolin 1'),\n",
       " ('ACSF2', 'acyl-CoA synthetase family member 2'),\n",
       " ('SIPA1', 'signal-induced proliferation-associated 1'),\n",
       " ('XRCC4',\n",
       "  'X-ray repair complementing defective repair in Chinese hamster cells 4'),\n",
       " ('FASLG', 'Fas ligand'),\n",
       " ('SLCO1B1', 'solute carrier organic anion transporter family member 1B1'),\n",
       " ('BTN2A1', 'butyrophilin subfamily 2 member A1'),\n",
       " ('LGALS3', 'lectin, galactoside-binding, soluble, 3'),\n",
       " ('GSTA4', 'glutathione S-transferase alpha 4'),\n",
       " ('SLC6A1', 'solute carrier family 6 member 1'),\n",
       " ('UGT1A1', 'UDP glucuronosyltransferase 1 family, polypeptide A1'),\n",
       " ('SLC6A4',\n",
       "  'solute carrier family 6 (neurotransmitter transporter), member 4'),\n",
       " ('SLC6A7', 'solute carrier family 6 member 7'),\n",
       " ('CACNA2D4', 'calcium voltage-gated channel auxiliary subunit alpha2delta 4'),\n",
       " ('ERLIN2', 'ER lipid raft associated 2'),\n",
       " ('SLC10A1', 'solute carrier family 10 member 1'),\n",
       " ('CD7', 'CD7 molecule'),\n",
       " ('RNF2', 'ring finger protein 2'),\n",
       " ('RDH16', 'retinol dehydrogenase 16 (all-trans)'),\n",
       " ('RBMX2', 'RNA binding motif protein, X-linked 2'),\n",
       " ('BRD4', 'bromodomain containing 4'),\n",
       " ('NUS1', 'NUS1 dehydrodolichyl diphosphate synthase subunit'),\n",
       " ('ABCB11', 'ATP binding cassette subfamily B member 11'),\n",
       " ('PRKCA', 'protein kinase C alpha'),\n",
       " ('HSH2D', 'hematopoietic SH2 domain containing'),\n",
       " ('LSS', 'lanosterol synthase (2,3-oxidosqualene-lanosterol cyclase)'),\n",
       " ('ADAM2', 'ADAM metallopeptidase domain 2'),\n",
       " ('PRKCZ', 'protein kinase C zeta'),\n",
       " ('SNCG', 'synuclein gamma'),\n",
       " ('ATP1B3', 'ATPase Na+/K+ transporting subunit beta 3'),\n",
       " ('ATG4D', 'autophagy related 4D, cysteine peptidase'),\n",
       " ('CDH8', 'cadherin 8'),\n",
       " ('MVK', 'mevalonate kinase'),\n",
       " ('GABRA2', 'gamma-aminobutyric acid type A receptor alpha2 subunit'),\n",
       " ('GABRA3', 'gamma-aminobutyric acid type A receptor alpha3 subunit'),\n",
       " ('GABRA5', 'gamma-aminobutyric acid type A receptor alpha5 subunit')]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[geneid2info[g] for g in geneids_study if g in geneid2info]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODO: Separate out metabolic genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "gg = goea_results_sig[0]\n",
    "gg.GO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://github.com/tanghaibao/goatools/blob/master/notebooks/goea_nbt3102.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    8 usr  40 GOs  WROTE: results3/out/go_enrich.png\n"
     ]
    }
   ],
   "source": [
    "from goatools.godag_plot import plot_gos, plot_results, plot_goid2goobj\n",
    "\n",
    "\n",
    "# Plot subset starting from these significant GO terms\n",
    "goid_subset = [g.GO for g in goea_results_sig] # [\n",
    "#     'GO:0003723', # MF D04 RNA binding (32 genes)\n",
    "#     'GO:0044822', # MF D05 poly(A) RNA binding (86 genes)\n",
    "#     'GO:0003729', # MF D06 mRNA binding (11 genes)\n",
    "#     'GO:0019843', # MF D05 rRNA binding (6 genes)\n",
    "#     'GO:0003746', # MF D06 translation elongation factor activity (5 genes)\n",
    "# ]\n",
    "plot_gos(\"results3/out/go_enrich.png\", \n",
    "    goid_subset, # Source GO ids\n",
    "    obodag, \n",
    "    goea_results=goea_results_all) # Use pvals for coloring\n",
    "# plot_results(\"nbt3102_BP.png\", goea_results_sig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This plot contains GOEA results:\n",
    "\n",
    "- GO terms colored by P-value:\n",
    "    - pval < 0.005 (light red)\n",
    "    - pval < 0.01 (light orange)\n",
    "    - pval < 0.05 (yellow)\n",
    "    - pval > 0.05 (grey) Study terms that are not statistically significant\n",
    "- GO terms with study gene counts printed. e.g., \"32 genes\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    8 usr  40 GOs  WROTE: results3/out/go_enrich_symbols.pdf\n"
     ]
    }
   ],
   "source": [
    "plot_gos(\"results3/out/go_enrich_symbols.pdf\", \n",
    "    goid_subset, # Source GO ids\n",
    "    obodag,\n",
    "    goea_results=goea_results_all, # use pvals for coloring\n",
    "    # We can further configure the plot...\n",
    "    id2symbol=geneid2symbol, # Print study gene Symbols, not Entrez GeneIDs\n",
    "    study_items=6, # Only only 6 gene Symbols max on GO terms\n",
    "    items_p_line=3, # Print 3 genes per line\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3-(4-methylsulfonylphenyl)-4-phenyl-2H-furan-5-one'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c = Compound.from_cid(5090)\n",
    "c.iupac_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "cids = [int(mapping.drug_idx_to_id[c].replace('CID','')) for c in data['pd_idx'][1]]\n",
    "drugs = [Compound.from_cid(int(cid)) for cid in set(cids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5732"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugs[0].cid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Compound(5732),\n",
       " Compound(4170),\n",
       " Compound(3002190),\n",
       " Compound(5039),\n",
       " Compound(2162),\n",
       " Compound(3325)]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drugs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5732"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dd = drugs[0]\n",
    "dd.cid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'CID152946'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-48-2c663c3a0819>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mmapping\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrug_id_to_idx\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'CID152946'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m: 'CID152946'"
     ]
    }
   ],
   "source": [
    "mapping.drug_id_to_idx['CID152946']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(104,\n",
       "  5732,\n",
       "  'zolpidem',\n",
       "  'N,N-dimethyl-2-[6-methyl-2-(4-methylphenyl)imidazo[1,2-a]pyridin-3-yl]acetamide'),\n",
       " (155,\n",
       "  4170,\n",
       "  'metolazone',\n",
       "  '7-chloro-2-methyl-3-(2-methylphenyl)-4-oxo-1,2-dihydroquinazoline-6-sulfonamide'),\n",
       " (172,\n",
       "  3002190,\n",
       "  'Telithromycin',\n",
       "  '(1S,2R,5R,7R,8R,9R,11R,13R,14R)-8-[(2S,3R,4S,6R)-4-(dimethylamino)-3-hydroxy-6-methyloxan-2-yl]oxy-2-ethyl-9-methoxy-1,5,7,9,11,13-hexamethyl-15-[4-(4-pyridin-3-ylimidazol-1-yl)butyl]-3,17-dioxa-15-azabicyclo[12.3.0]heptadecane-4,6,12,16-tetrone'),\n",
       " (31,\n",
       "  5039,\n",
       "  'UNII-884KT10YB7',\n",
       "  \"1-N'-[2-[[5-[(dimethylamino)methyl]furan-2-yl]methylsulfanyl]ethyl]-1-N-methyl-2-nitroethene-1,1-diamine\"),\n",
       " (34,\n",
       "  2162,\n",
       "  'amlodipine',\n",
       "  '3-O-ethyl 5-O-methyl 2-(2-aminoethoxymethyl)-4-(2-chlorophenyl)-6-methyl-1,4-dihydropyridine-3,5-dicarboxylate'),\n",
       " (96,\n",
       "  3325,\n",
       "  'famotidine',\n",
       "  \"3-[[2-(diaminomethylideneamino)-1,3-thiazol-4-yl]methylsulfanyl]-N'-sulfamoylpropanimidamide\")]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_ids = [(mapping.drug_id_to_idx['CID{}'.format(d.cid)], d.cid, d.synonyms[0], d.iupac_name) for d in drugs]\n",
    "drug_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "fps = [d.fingerprint for d in drugs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ABCB1'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneid2symbol[5243]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Zielinski et al. (2015) Nat Commun"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "147\n",
      "24\n"
     ]
    }
   ],
   "source": [
    "#mapping.drug_id_to_idx['CID3559'] # haloperidol\n",
    "#print(mapping.drug_id_to_idx['CID4091']) # metformin--doesn't exist\n",
    "print(mapping.drug_id_to_idx['CID2801']) # clomipramine\n",
    "print(mapping.drug_id_to_idx['CID5523']) # Ultram "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "import binascii\n",
    "bs = binascii.unhexlify(fps[0])\n",
    "bs\n",
    "\n",
    "from scipy.spatial import distance\n",
    "\n",
    "bfps = [binascii.unhexlify(f) for f in fps]\n",
    "# distance.rogerstanimoto(bfps[0],bfps[2])\n",
    "\n",
    "bb = bfps[0]\n",
    "bb.\n",
    "\n",
    "bfps[0]\n",
    "\n",
    "bfps[2]\n",
    "\n",
    "# Data will be stored in this variable\n",
    "import os\n",
    "geneid2symbol = {}\n",
    "# Get xlsx filename where data is stored\n",
    "ROOT = os.path.dirname(os.getcwd()) # go up 1 level from current working directory\n",
    "din_xlsx = \"../data/nbt.3102-S4_GeneIDs.xlsx\"\n",
    "# Read data\n",
    "if os.path.isfile(din_xlsx):  \n",
    "    import xlrd\n",
    "    book = xlrd.open_workbook(din_xlsx)\n",
    "    pg = book.sheet_by_index(0)\n",
    "    for r in range(pg.nrows):\n",
    "        symbol, geneid, pval = [pg.cell_value(r, c) for c in range(pg.ncols)]\n",
    "        if geneid:\n",
    "            geneid2symbol[int(geneid)] = symbol\n",
    "    print('{N} genes READ: {XLSX}'.format(N=len(geneid2symbol), XLSX=din_xlsx))\n",
    "else:\n",
    "    raise RuntimeError('FILE NOT FOUND: {XLSX}'.format(XLSX=din_xlsx))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
