{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Paper next results\n",
    "- validate link between arrhythmia, myopathy, and the identified genes (SLCO1B1)\n",
    "- validate the drugs involved in arrhythmia\n",
    "    - run PoSe-Path on the drugs found in literature for arrhythmia; check for other side effects (and check in literature)\n",
    "- validate the other promising heart side effects\n",
    "    - explain (jointly) related side effects (having similar mechanisms) in the heart and across organs (e.g., heart muscle, other muscle\n",
    "- validate muscle (myopathy) since connected with cardiovascular disease\n",
    "- move on to validating liver (since SLCO1B1 is in liver), kidney, (brain)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Explainer paper\n",
    "1. Each single side effect --> Enriched GO --> validate with literature\n",
    "1. Choose two side effects suspected to share mechanisms --> get common subgraph --> validate\n",
    "1. Choose >2 side effects ...\n",
    "1. Choose single interesting drug --> Enriched GO --> validate with literature\n",
    "1. Choose interesting drug-drug interaction --> Enriched GO --> validate with literature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import goatools\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import pubchempy\n",
    "import pickle\n",
    "\n",
    "from pubchempy import Compound\n",
    "\n",
    "# Get http://geneontology.org/ontology/go-basic.obo\n",
    "from goatools.base import download_go_basic_obo\n",
    "# Get ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene2go.gz\n",
    "from goatools.base import download_ncbi_associations\n",
    "from goatools.obo_parser import GODag\n",
    "from __future__ import print_function\n",
    "from goatools.anno.genetogo_reader import Gene2GoReader\n",
    "from goatools.test_data.genes_NCBI_9606_ProteinCoding import GENEID2NT as GeneID2nt_hum\n",
    "from goatools.goea.go_enrichment_ns import GOEnrichmentStudyNS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_colwidth', -1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add the Side Effect  Disease Class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_se = pd.read_csv('data/Se-DoDecagon_sidefx.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(561, 3)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_se.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Side Effect</th>\n",
       "      <th>Side Effect Name</th>\n",
       "      <th>Disease Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C0017152</td>\n",
       "      <td>gastric inflammation</td>\n",
       "      <td>gastrointestinal system disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C0027858</td>\n",
       "      <td>neuroma</td>\n",
       "      <td>benign neoplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C0041466</td>\n",
       "      <td>Typhoid</td>\n",
       "      <td>bacterial infectious disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C0032807</td>\n",
       "      <td>Post thrombotic syndrome</td>\n",
       "      <td>cardiovascular system disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C0033860</td>\n",
       "      <td>psoriasis</td>\n",
       "      <td>integumentary system disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Side Effect          Side Effect Name                    Disease Class\n",
       "0  C0017152    gastric inflammation      gastrointestinal system disease\n",
       "1  C0027858    neuroma                   benign neoplasm                \n",
       "2  C0041466    Typhoid                   bacterial infectious disease   \n",
       "3  C0032807    Post thrombotic syndrome  cardiovascular system disease  \n",
       "4  C0033860    psoriasis                 integumentary system disease   "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_se.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['gastrointestinal system disease', 'benign neoplasm',\n",
       "       'bacterial infectious disease', 'cardiovascular system disease',\n",
       "       'integumentary system disease', 'nervous system disease',\n",
       "       'cognitive disorder', 'urinary system disease',\n",
       "       'hematopoietic system disease', 'endocrine system disease',\n",
       "       'musculoskeletal system disease', 'psoriatic arthritis',\n",
       "       'hematopoietic system diseases', 'viral infectious disease',\n",
       "       'fungal infectious disease', 'respiratory system disease',\n",
       "       'cancer', 'developmental disorder of mental health',\n",
       "       'acquired metabolic disease', 'parasitic infectious disease',\n",
       "       'thoracic disease', 'inherited metabolic disorder',\n",
       "       'reproductive system disease', 'immune system disease',\n",
       "       'monogenic disease', 'substance-related disorder',\n",
       "       'personality disorder', 'hypospadias', 'sleep disorder',\n",
       "       'sexual disorder', 'chromosomal disease',\n",
       "       'polycystic ovary syndrome', 'somatoform disorder',\n",
       "       'pre-malignant neoplasm', 'orofacial cleft', 'cryptorchidism',\n",
       "       'irritable bowel syndrome'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_se['Disease Class'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/laurence/git/posepath/PoSe-Path-3.0/data/tipexp_data.pkl','rb') as f:\n",
    "    mapping = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get disease class for SE where available"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(861, 4)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_se0 = pd.DataFrame([{'Side Effect Name':k.lower(), 'idx':v} for k,v in mapping.side_effect_name_to_idx.items()])\n",
    "df_se.loc[:,'Side Effect Name'] = df_se['Side Effect Name'].str.lower()\n",
    "df_ses = pd.merge(df_se0, df_se, on='Side Effect Name', how='left')\n",
    "df_ses.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Side Effect Name</th>\n",
       "      <th>idx</th>\n",
       "      <th>Side Effect</th>\n",
       "      <th>Disease Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>atelectasis</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>back ache</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>agitated</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>abnormal movements</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>acidosis</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>apnea</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>drug hypersensitivity</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>flatulence</td>\n",
       "      <td>9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pain in throat</td>\n",
       "      <td>10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>bradycardia</td>\n",
       "      <td>13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>lung infiltration</td>\n",
       "      <td>14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>bleeding</td>\n",
       "      <td>15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>hypoglycaemia neonatal</td>\n",
       "      <td>16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>peritonitis</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>abdominal distension</td>\n",
       "      <td>20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>hypoxia</td>\n",
       "      <td>23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>difficulty breathing</td>\n",
       "      <td>24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>decreased body temperature</td>\n",
       "      <td>25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>sepsis</td>\n",
       "      <td>27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>hypomagnesaemia</td>\n",
       "      <td>28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>appendectomy</td>\n",
       "      <td>31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>flank pain</td>\n",
       "      <td>32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>pneumothorax</td>\n",
       "      <td>33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>strabismus</td>\n",
       "      <td>35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>blood calcium decreased</td>\n",
       "      <td>36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>respiratory failure</td>\n",
       "      <td>37</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>hypercapnia</td>\n",
       "      <td>39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>leucocytosis</td>\n",
       "      <td>41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>hot flash</td>\n",
       "      <td>43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>arterial pressure nos decreased</td>\n",
       "      <td>44</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>811</th>\n",
       "      <td>kidney transplant</td>\n",
       "      <td>811</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>813</th>\n",
       "      <td>fracture nonunion</td>\n",
       "      <td>813</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>815</th>\n",
       "      <td>empyema</td>\n",
       "      <td>815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>gastric ulcer perforated</td>\n",
       "      <td>816</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>822</th>\n",
       "      <td>neurogenic bladder</td>\n",
       "      <td>822</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>823</th>\n",
       "      <td>uterine bleeding</td>\n",
       "      <td>823</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>uterine polyp</td>\n",
       "      <td>824</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>hyperacusia</td>\n",
       "      <td>825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>826</th>\n",
       "      <td>enterocolitis</td>\n",
       "      <td>826</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>motor retardation</td>\n",
       "      <td>827</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828</th>\n",
       "      <td>cryptococcosis</td>\n",
       "      <td>828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830</th>\n",
       "      <td>food intolerance</td>\n",
       "      <td>830</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>831</th>\n",
       "      <td>cluster headache</td>\n",
       "      <td>831</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>833</th>\n",
       "      <td>macrocytosis</td>\n",
       "      <td>833</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>834</th>\n",
       "      <td>anisocoria</td>\n",
       "      <td>834</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>836</th>\n",
       "      <td>menstrual disorder</td>\n",
       "      <td>836</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>838</th>\n",
       "      <td>irregular menstrual cycle</td>\n",
       "      <td>838</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>839</th>\n",
       "      <td>pancreatic pseudocyst</td>\n",
       "      <td>839</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>840</th>\n",
       "      <td>testicular swelling</td>\n",
       "      <td>840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>843</th>\n",
       "      <td>sacroiliitis</td>\n",
       "      <td>843</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>tracheobronchitis</td>\n",
       "      <td>844</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847</th>\n",
       "      <td>xerosis</td>\n",
       "      <td>847</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>848</th>\n",
       "      <td>postmenopausal bleeding</td>\n",
       "      <td>848</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>852</th>\n",
       "      <td>testes pain</td>\n",
       "      <td>852</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>853</th>\n",
       "      <td>esophageal stenosis</td>\n",
       "      <td>853</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>854</th>\n",
       "      <td>coccydynia</td>\n",
       "      <td>854</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>855</th>\n",
       "      <td>corneal abrasion</td>\n",
       "      <td>855</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>856</th>\n",
       "      <td>bunion</td>\n",
       "      <td>856</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>857</th>\n",
       "      <td>viral pharyngitis</td>\n",
       "      <td>857</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>860</th>\n",
       "      <td>endometriosis</td>\n",
       "      <td>860</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>521 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Side Effect Name  idx Side Effect Disease Class\n",
       "0    atelectasis         0    NaN         NaN         \n",
       "1    back ache           1    NaN         NaN         \n",
       "3    agitated            3    NaN         NaN         \n",
       "4    abnormal movements  4    NaN         NaN         \n",
       "5    acidosis            5    NaN         NaN         \n",
       "..        ...           ..    ...         ...         \n",
       "854  coccydynia          854  NaN         NaN         \n",
       "855  corneal abrasion    855  NaN         NaN         \n",
       "856  bunion              856  NaN         NaN         \n",
       "857  viral pharyngitis   857  NaN         NaN         \n",
       "860  endometriosis       860  NaN         NaN         \n",
       "\n",
       "[521 rows x 4 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ses[df_ses['Disease Class'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Side Effect Name</th>\n",
       "      <th>idx</th>\n",
       "      <th>Side Effect</th>\n",
       "      <th>Disease Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>acute myeloblastic leukemia</td>\n",
       "      <td>390</td>\n",
       "      <td>C0023467</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>adenocarcinoma</td>\n",
       "      <td>61</td>\n",
       "      <td>C0001418</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>basal cell carcinoma</td>\n",
       "      <td>574</td>\n",
       "      <td>C0007117</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>bladder cancer</td>\n",
       "      <td>799</td>\n",
       "      <td>C0699885</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>672</th>\n",
       "      <td>brain neoplasm</td>\n",
       "      <td>672</td>\n",
       "      <td>C0006118</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>breast cancer</td>\n",
       "      <td>85</td>\n",
       "      <td>C0678222</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>cancer</td>\n",
       "      <td>617</td>\n",
       "      <td>C0006826</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>carcinoma of prostate</td>\n",
       "      <td>688</td>\n",
       "      <td>C0600139</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558</th>\n",
       "      <td>carcinoma of the colon</td>\n",
       "      <td>558</td>\n",
       "      <td>C0699790</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821</th>\n",
       "      <td>esophageal cancer</td>\n",
       "      <td>821</td>\n",
       "      <td>C0152018</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>766</th>\n",
       "      <td>gastric cancer</td>\n",
       "      <td>766</td>\n",
       "      <td>C0699791</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>hepatic neoplasia</td>\n",
       "      <td>187</td>\n",
       "      <td>C0023903</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>leukaemia</td>\n",
       "      <td>142</td>\n",
       "      <td>C0023418</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>lymphoma</td>\n",
       "      <td>599</td>\n",
       "      <td>C0024299</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>malignant melanoma</td>\n",
       "      <td>483</td>\n",
       "      <td>C0025202</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>781</th>\n",
       "      <td>meningioma</td>\n",
       "      <td>781</td>\n",
       "      <td>C0025286</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>551</th>\n",
       "      <td>multiple myeloma</td>\n",
       "      <td>551</td>\n",
       "      <td>C0026764</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>pancreatic cancer</td>\n",
       "      <td>648</td>\n",
       "      <td>C0235974</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>squamous cell carcinoma of skin</td>\n",
       "      <td>647</td>\n",
       "      <td>C0553723</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>372</th>\n",
       "      <td>thyroid neoplasia</td>\n",
       "      <td>372</td>\n",
       "      <td>C0040136</td>\n",
       "      <td>cancer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Side Effect Name  idx Side Effect Disease Class\n",
       "390  acute myeloblastic leukemia      390  C0023467    cancer      \n",
       "61   adenocarcinoma                   61   C0001418    cancer      \n",
       "574  basal cell carcinoma             574  C0007117    cancer      \n",
       "799  bladder cancer                   799  C0699885    cancer      \n",
       "672  brain neoplasm                   672  C0006118    cancer      \n",
       "85   breast cancer                    85   C0678222    cancer      \n",
       "617  cancer                           617  C0006826    cancer      \n",
       "688  carcinoma of prostate            688  C0600139    cancer      \n",
       "558  carcinoma of the colon           558  C0699790    cancer      \n",
       "821  esophageal cancer                821  C0152018    cancer      \n",
       "766  gastric cancer                   766  C0699791    cancer      \n",
       "187  hepatic neoplasia                187  C0023903    cancer      \n",
       "142  leukaemia                        142  C0023418    cancer      \n",
       "599  lymphoma                         599  C0024299    cancer      \n",
       "483  malignant melanoma               483  C0025202    cancer      \n",
       "781  meningioma                       781  C0025286    cancer      \n",
       "551  multiple myeloma                 551  C0026764    cancer      \n",
       "648  pancreatic cancer                648  C0235974    cancer      \n",
       "647  squamous cell carcinoma of skin  647  C0553723    cancer      \n",
       "372  thyroid neoplasia                372  C0040136    cancer      "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ses.dropna()[ df_ses['Disease Class'].dropna().str.contains('cancer')].sort_values('Side Effect Name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Side Effect Name</th>\n",
       "      <th>idx</th>\n",
       "      <th>Side Effect</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Disease Class</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>nervous system disease</th>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cardiovascular system disease</th>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gastrointestinal system disease</th>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>musculoskeletal system disease</th>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>integumentary system disease</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>respiratory system disease</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cancer</th>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>urinary system disease</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hematopoietic system disease</th>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>acquired metabolic disease</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>endocrine system disease</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>immune system disease</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reproductive system disease</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>viral infectious disease</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fungal infectious disease</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cognitive disorder</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>benign neoplasm</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>inherited metabolic disorder</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>developmental disorder of mental health</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>thoracic disease</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bacterial infectious disease</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hematopoietic system diseases</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>personality disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pre-malignant neoplasm</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>psoriatic arthritis</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sexual disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sleep disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>somatoform disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>substance-related disorder</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>irritable bowel syndrome</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Side Effect Name  idx  Side Effect\n",
       "Disease Class                                                              \n",
       "nervous system disease                   46                46   46         \n",
       "cardiovascular system disease            45                45   45         \n",
       "gastrointestinal system disease          41                41   41         \n",
       "musculoskeletal system disease           27                27   27         \n",
       "integumentary system disease             21                21   21         \n",
       "respiratory system disease               21                21   21         \n",
       "cancer                                   20                20   20         \n",
       "urinary system disease                   19                19   19         \n",
       "hematopoietic system disease             15                15   15         \n",
       "acquired metabolic disease               12                12   12         \n",
       "endocrine system disease                 12                12   12         \n",
       "immune system disease                    9                 9    9          \n",
       "reproductive system disease              7                 7    7          \n",
       "viral infectious disease                 7                 7    7          \n",
       "fungal infectious disease                7                 7    7          \n",
       "cognitive disorder                       6                 6    6          \n",
       "benign neoplasm                          6                 6    6          \n",
       "inherited metabolic disorder             3                 3    3          \n",
       "developmental disorder of mental health  3                 3    3          \n",
       "thoracic disease                         3                 3    3          \n",
       "bacterial infectious disease             1                 1    1          \n",
       "hematopoietic system diseases            1                 1    1          \n",
       "personality disorder                     1                 1    1          \n",
       "pre-malignant neoplasm                   1                 1    1          \n",
       "psoriatic arthritis                      1                 1    1          \n",
       "sexual disorder                          1                 1    1          \n",
       "sleep disorder                           1                 1    1          \n",
       "somatoform disorder                      1                 1    1          \n",
       "substance-related disorder               1                 1    1          \n",
       "irritable bowel syndrome                 1                 1    1          "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ses.groupby('Disease Class').count().sort_values('Side Effect', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "508"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping.side_effect_name_to_idx['Arrhythmia']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "basepath = 'results/results/out/all-all-508-3.0-0.97'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#with open('../data/[91.0, 88.0]-[84.0, 84.0]-[28.0, 28.0]-1.0.pkl','rb') as f:\n",
    "#with open('results/out/pkl/all-all-28-1.0-0.99.pkl','rb') as f:\n",
    "#with open('results2/out/pkl/100,34,88-100,34,88-28-1.0-0.5.pkl','rb') as f:\n",
    "#with open('results3/out/pkl/147-all-all-3.0-0.96.pkl','rb') as f:\n",
    "#with open('results3/out/pkl/all-all-508-3.0-0.97.pkl','rb') as f:  # previous\n",
    "#with open('results/results/out/all-all-536-3.0-0.97/exp_info.pkl','rb') as f:\n",
    "with open(os.path.join(basepath, 'exp_info.pkl'),'rb') as f:\n",
    "    data = pickle.load(f)\n",
    "dmap = data.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "idx1 = data['pp_idx'][0]\n",
    "idx2 = data['pp_idx'][1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['pp_id'] = [[int(mapping.prot_idx_to_id[idx].replace('GeneID','')) for idx in idx1],\n",
    "                 [int(mapping.prot_idx_to_id[idx].replace('GeneID','')) for idx in idx2]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check the namespace directly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_go = pd.read_csv(os.path.join(basepath, 'name_space.csv'), index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['biological_process', 'cellular_component', 'molecular_function'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go.namespace.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>bile acid transmembrane transporter activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.006280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>sodium-independent organic anion transmembrane transporter activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.006849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ATPase-coupled transmembrane transporter activity</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>0.037601</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  name  \\\n",
       "5  bile acid transmembrane transporter activity                          \n",
       "6  sodium-independent organic anion transmembrane transporter activity   \n",
       "7  ATPase-coupled transmembrane transporter activity                     \n",
       "\n",
       "            namespace  p_fdr_bh  \n",
       "5  molecular_function  0.006280  \n",
       "6  molecular_function  0.006849  \n",
       "7  molecular_function  0.037601  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go[ df_go.namespace.str.contains('molecular')].sort_values('p_fdr_bh')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Example 2: Clomipramine --> side effect --> PPI\n",
    "- Drug 1: 147 (Clomipramine)\n",
    "- Drug 2: 24 (all-->one)\n",
    "- Side effect: 376 (aortic regurgitation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze after running below"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "set([mapping.side_effect_idx_to_name[k] for k in data['sd_idx']])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "d2 = [mapping.drug_idx_to_id[k] for k in data['durg2']]\n",
    "\n",
    "cids = [int(d.replace('CID','')) for d in d2]\n",
    "drugs = [Compound.from_cid(int(cid)) for cid in set(cids)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "drugs\n",
    "\n",
    "drug_ids = [(d.cid, d.synonyms[0], d.iupac_name) for d in drugs]\n",
    "drug_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the side effects to run PoSe-Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "hearts = ['arrhythmia','cardiac','heart','ventric','atri','aort']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('heart attack', 186),\n",
       " ('arteriosclerotic heart disease', 202),\n",
       " ('Extrasystoles Ventricular', 239),\n",
       " ('Atrioventricular block first degree', 259),\n",
       " ('Cardiac decompensation', 287),\n",
       " ('heart rate increased', 327),\n",
       " ('aortic regurgitation', 358),\n",
       " ('cardiac murmur', 398),\n",
       " ('cardiac enlargement', 399),\n",
       " ('cardiac disease', 448),\n",
       " ('atrial septal defect', 453),\n",
       " ('atrial ectopic beats', 486),\n",
       " ('Arrhythmia', 508),\n",
       " ('Cardiac ischemia', 516),\n",
       " ('cardiac failure', 527),\n",
       " ('atrioventricular block second degree', 528),\n",
       " ('tachycardia ventricular', 535),\n",
       " ('atrial flutter', 538),\n",
       " ('cardiac valvulopathy', 560),\n",
       " ('Atrioventricular block complete', 562),\n",
       " ('ventricular fibrillation', 563),\n",
       " ('Cardiac tamponade', 580),\n",
       " ('Supraventricular tachycardia', 594),\n",
       " ('right heart failure', 615),\n",
       " ('aortic stenosis', 642),\n",
       " ('atrioventricular block', 656),\n",
       " ('congenital heart disease', 686),\n",
       " ('aortic aneurysm', 740),\n",
       " ('left ventricular hypertrophy', 773),\n",
       " ('Bradyarrhythmia', 791)]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ses = [(k,v) for k,v in mapping.side_effect_name_to_idx.items() \n",
    " if any([x.lower() in k.lower() for x in hearts])]\n",
    "ses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mapping.side_effect_idx_to_name)\n",
    "name_to_idx = {v:k for k, v in mapping.side_effect_idx_to_name.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'186,202,239,259,287,327,358,398,399,448,453,486,508,516,527,528,535,538,560,562,563,580,594,615,642,656,686,740,773,791'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "','.join([str(x[1]) for x in ses])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Chronic Kidney Disease', 143),\n",
       " ('acute kidney failure', 144),\n",
       " ('kidney failure', 256),\n",
       " ('renal cyst', 277),\n",
       " ('adrenal insufficiency', 301),\n",
       " ('disorder Renal', 333),\n",
       " ('renal tubular acidosis', 501),\n",
       " ('renal mass', 690),\n",
       " ('hepatorenal syndrome', 711),\n",
       " ('kidney pain', 730),\n",
       " ('kidney transplant', 811)]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "renal = ['kidne','renal']\n",
    "ses = [(k,v) for k,v in mapping.side_effect_name_to_idx.items() \n",
    " if any([x.lower() in k.lower() for x in renal])]\n",
    "ses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('muscle spasm', 126),\n",
       " ('muscle paresis', 137),\n",
       " ('aching muscles', 175),\n",
       " ('muscle weakness', 190),\n",
       " ('Cardiomyopathy', 283),\n",
       " ('muscle strain', 416),\n",
       " ('muscle inflammation', 488),\n",
       " ('muscle disorder', 533)]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muscle = ['muscl','myopath']\n",
    "ses = [(k,v) for k,v in mapping.side_effect_name_to_idx.items() \n",
    " if any([x.lower() in k.lower() for x in muscle])]\n",
    "ses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Muscle / Heart / transporter genes\n",
    "https://ascpt.onlinelibrary.wiley.com/doi/abs/10.1038/clpt.2013.234\n",
    "\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pubmed/29463526\n",
    "- potentially strong assocation between arrythmia and SLCO1B1 due to increased risk of prolonged QTc interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GO enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneid2symbol = {v.GeneID: v.Symbol for k,v in GeneID2nt_hum.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2020-01-01) 47,337 GO Terms\n",
      "HMS:0:00:02.921323 323,107 annotations READ: gene2go \n",
      "1 taxids stored: 9606\n",
      "BP 17,541 annotated human genes\n",
      "MF 17,384 annotated human genes\n",
      "CC 18,648 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 80% 16,711 of 20,913 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 85% 17,755 of 20,913 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 80% 16,699 of 20,913 population items found in association\n"
     ]
    }
   ],
   "source": [
    "\n",
    "obo_fname = download_go_basic_obo()\n",
    "\n",
    "\n",
    "fin_gene2go = download_ncbi_associations()\n",
    "\n",
    "obodag = GODag(\"go-basic.obo\")\n",
    "\n",
    "# Read NCBI's gene2go. Store annotations in a list of namedtuples\n",
    "objanno = Gene2GoReader(fin_gene2go, taxids=[9606])\n",
    "\n",
    "# Get namespace2association where:\n",
    "#    namespace is:\n",
    "#        BP: biological_process               \n",
    "#        MF: molecular_function\n",
    "#        CC: cellular_component\n",
    "#    assocation is a dict:\n",
    "#        key: NCBI GeneID\n",
    "#        value: A set of GO IDs associated with that gene\n",
    "ns2assoc = objanno.get_ns2assc()\n",
    "\n",
    "for nspc, id2gos in ns2assoc.items():\n",
    "    print(\"{NS} {N:,} annotated human genes\".format(NS=nspc, N=len(id2gos)))\n",
    "\n",
    "\n",
    "goeaobj = GOEnrichmentStudyNS(\n",
    "        GeneID2nt_hum.keys(), # List of human protein-acoding genes\n",
    "        ns2assoc, # geneid/GO associations\n",
    "        obodag, # Ontologies\n",
    "        propagate_counts = False,\n",
    "        alpha = 0.05, # default significance cut-off\n",
    "        methods = ['fdr_bh']) # defult multipletest correction method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 4 IDs ...\n",
      "100%      4 of      4 study items found in association\n",
      "100%      4 of      4 study items found in population(20913)\n",
      "Calculating 12,189 uncorrected p-values using fisher_scipy_stats\n",
      "  12,189 GO terms are associated with 16,711 of 20,913 population items\n",
      "      18 GO terms are associated with      4 of      4 study items\n",
      "  METHOD fdr_bh:\n",
      "       4 GO terms found significant (< 0.05=alpha) (  4 enriched +   0 purified): statsmodels fdr_bh\n",
      "       4 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 4 IDs ...\n",
      "100%      4 of      4 study items found in association\n",
      "100%      4 of      4 study items found in population(20913)\n",
      "Calculating 1,731 uncorrected p-values using fisher_scipy_stats\n",
      "   1,731 GO terms are associated with 17,755 of 20,913 population items\n",
      "      12 GO terms are associated with      4 of      4 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "       3 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 4 IDs ...\n",
      "100%      4 of      4 study items found in association\n",
      "100%      4 of      4 study items found in population(20913)\n",
      "Calculating 4,164 uncorrected p-values using fisher_scipy_stats\n",
      "   4,164 GO terms are associated with 16,699 of 20,913 population items\n",
      "      15 GO terms are associated with      4 of      4 study items\n",
      "  METHOD fdr_bh:\n",
      "       3 GO terms found significant (< 0.05=alpha) (  3 enriched +   0 purified): statsmodels fdr_bh\n",
      "       4 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n"
     ]
    }
   ],
   "source": [
    "# 'p_' means \"pvalue\". 'fdr_bh' is the multipletest method we are currently using.\n",
    "# geneids_study = data['pp_id'][0] + data['pp_id'][1] # geneid2symbol.keys()\n",
    "geneids_study = set(data['pp_id'][0] + data['pp_id'][1])\n",
    "goea_results_all = goeaobj.run_study(geneids_study)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "goea_results_sig = [r for r in goea_results_all if r.p_fdr_bh < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(goea_results_sig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{8647, 10599, 28234}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0 = goea_results_sig[0]\n",
    "g0.study_items"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bile acid and bile salt transport</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>0.000105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>transmembrane transport</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>0.000200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>organic anion transport</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>0.010025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sodium-independent organic anion transport</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>0.010025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>basolateral plasma membrane</td>\n",
       "      <td>cellular_component</td>\n",
       "      <td>0.005575</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         name           namespace  p_fdr_bh\n",
       "0  bile acid and bile salt transport           biological_process  0.000105\n",
       "1  transmembrane transport                     biological_process  0.000200\n",
       "2  organic anion transport                     biological_process  0.010025\n",
       "3  sodium-independent organic anion transport  biological_process  0.010025\n",
       "4  basolateral plasma membrane                 cellular_component  0.005575"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GOTerm('GO:0015721'):\n",
       "  id:GO:0015721\n",
       "  item_id:GO:0015721\n",
       "  name:bile acid and bile salt transport\n",
       "  namespace:biological_process\n",
       "  _parents: 3 items\n",
       "    GO:0015850\n",
       "    GO:0006869\n",
       "    GO:0015718\n",
       "  parents: 3 items\n",
       "    GO:0006869\tlevel-05\tdepth-05\tlipid transport [biological_process]\n",
       "    GO:0015850\tlevel-05\tdepth-05\torganic hydroxy compound transport [biological_process]\n",
       "    GO:0015718\tlevel-07\tdepth-08\tmonocarboxylic acid transport [biological_process]\n",
       "  children: 1 items\n",
       "    GO:0015722\tlevel-06\tdepth-10\tcanalicular bile acid transport [biological_process]\n",
       "  level:6\n",
       "  depth:9\n",
       "  is_obsolete:False\n",
       "  alt_ids: 0 items"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0 = goea_results_sig[0]\n",
    "g0.goterm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'biological_process'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0.goterm.__dict__['namespace']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.00010470005154705395"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0.__dict__['p_fdr_bh']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GO:0015721'"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g0 = goea_results_sig[0]\n",
    "g0.goterm.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = ['name','namespace','id']\n",
    "df_go1 = pd.DataFrame([\n",
    "        {k:g.goterm.__dict__.get(k) for k in keys}\n",
    "        for g in goea_results_sig])\n",
    "df_p = pd.DataFrame([{'p_fdr_bh':g.__dict__['p_fdr_bh']} for g in goea_results_sig])\n",
    "df_go = df_go1.merge(df_p, left_index=True, right_index=True)\n",
    "go_genes = pd.DataFrame([{'id':g.goterm.id, 'gene':s,'symbol':geneid2symbol[s]} for g in goea_results_sig for s in g.study_items])\n",
    "df_go = df_go.merge(go_genes, on='id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "      <th>gene</th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>namespace</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>biological_process</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cellular_component</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>molecular_function</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    name  id  p_fdr_bh  gene  symbol\n",
       "namespace                                           \n",
       "biological_process  11    11  11        11    11    \n",
       "cellular_component  3     3   3         3     3     \n",
       "molecular_function  6     6   6         6     6     "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go.groupby('namespace').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "from cobra.io import load_json_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = load_json_model('/home/laurence/models/BiGG_M/json/Recon3D.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 755.003</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10FTHF5GLUtl</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF5GLUtm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF6GLUtl</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF6GLUtm</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10FTHF7GLUtl</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYOR_u10mi</th>\n",
       "      <td>433.920904</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Htmi</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NADH2_u10mi</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYOOm3i</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYOOm2i</th>\n",
       "      <td>566.079096</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10600 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 755.003 at 0x7f71ee57e3c8>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([28234, 10599,  8647, 10257])"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sig = df_go.gene.unique()\n",
    "genes_sig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_gene_ids = [int(g.id.split('_')[0].strip()) for g in model.genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_sig_M = [g for g in genes_sig if g in M_gene_ids]\n",
    "len(genes_sig_M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_g_rxn = pd.DataFrame([{'gene':int(g.id.split('_')[0].strip()),'rxn':r.reaction,\n",
    "                          'rxn_id':r.id, 'rxn_name':r.name, 'subsystem':r.subsystem}\n",
    "                         for g in model.genes for r in g.reactions])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>rxn</th>\n",
       "      <th>rxn_id</th>\n",
       "      <th>rxn_name</th>\n",
       "      <th>subsystem</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>hdca24g_c &lt;=&gt; hdca24g_r</td>\n",
       "      <td>HDCA24Gtr</td>\n",
       "      <td>Glucuronidated Compound Transport (HDCA-24G)</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8639</td>\n",
       "      <td>h2o_c + mma_c + o2_c --&gt; fald_c + h2o2_c + nh4_c</td>\n",
       "      <td>MAOX</td>\n",
       "      <td>Methylamine:oxygen oxidoreductase (deaminating) (copper-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8639</td>\n",
       "      <td>aact_c + h2o_c + o2_c --&gt; h2o2_c + mthgxl_c + nh4_c</td>\n",
       "      <td>AACTOOR</td>\n",
       "      <td>Aminoacetone:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Glycine, serine, alanine, and threonine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8639</td>\n",
       "      <td>h2o_c + o2_c + ptrc_c --&gt; 4abutn_c + h2o2_c + nh4_c</td>\n",
       "      <td>PTRCOX1</td>\n",
       "      <td>Putrescine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8639</td>\n",
       "      <td>dopa_c + h2o_c + o2_c --&gt; 34dhpac_c + h2o2_c + nh4_c</td>\n",
       "      <td>42A12BOOX</td>\n",
       "      <td>4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>8639</td>\n",
       "      <td>h2o_c + mhista_c + o2_c --&gt; 3mldz_c + h2o2_c + nh4_c</td>\n",
       "      <td>MHISOR</td>\n",
       "      <td>N-Methylhistamine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>8639</td>\n",
       "      <td>13dampp_c + h2o_c + o2_c --&gt; bamppald_c + h2o2_c + nh4_c</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>1,3-Diaminopropane:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Beta-Alanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>314</td>\n",
       "      <td>dopa_c + h2o_c + o2_c --&gt; 34dhpac_c + h2o2_c + nh4_c</td>\n",
       "      <td>42A12BOOX</td>\n",
       "      <td>4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + peamn_c --&gt; h2o2_c + nh4_c + pacald_c</td>\n",
       "      <td>PEAMNO</td>\n",
       "      <td>Phenethylamine oxidase</td>\n",
       "      <td>Phenylalanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + mma_c + o2_c --&gt; fald_c + h2o2_c + nh4_c</td>\n",
       "      <td>MAOX</td>\n",
       "      <td>Methylamine:oxygen oxidoreductase (deaminating) (copper-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + trypta_c --&gt; h2o2_c + id3acald_c + nh4_c</td>\n",
       "      <td>TRYPTAOX</td>\n",
       "      <td>Tryptamine:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tryptophan metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + hista_c + o2_c --&gt; h2o2_c + im4act_c + nh4_c</td>\n",
       "      <td>HISTASE</td>\n",
       "      <td>HISTASE</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + ptrc_c --&gt; 4abutn_c + h2o2_c + nh4_c</td>\n",
       "      <td>PTRCOX1</td>\n",
       "      <td>Putrescine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + mhista_c + o2_c --&gt; 3mldz_c + h2o2_c + nh4_c</td>\n",
       "      <td>MHISOR</td>\n",
       "      <td>N-Methylhistamine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>314</td>\n",
       "      <td>13dampp_c + h2o_c + o2_c --&gt; bamppald_c + h2o2_c + nh4_c</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>1,3-Diaminopropane:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Beta-Alanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>314</td>\n",
       "      <td>dopa_c + h2o_c + o2_c --&gt; 34dhpac_c + h2o2_c + nh4_c</td>\n",
       "      <td>42A12BOOX</td>\n",
       "      <td>4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + peamn_c --&gt; h2o2_c + nh4_c + pacald_c</td>\n",
       "      <td>PEAMNO</td>\n",
       "      <td>Phenethylamine oxidase</td>\n",
       "      <td>Phenylalanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + mma_c + o2_c --&gt; fald_c + h2o2_c + nh4_c</td>\n",
       "      <td>MAOX</td>\n",
       "      <td>Methylamine:oxygen oxidoreductase (deaminating) (copper-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + trypta_c --&gt; h2o2_c + id3acald_c + nh4_c</td>\n",
       "      <td>TRYPTAOX</td>\n",
       "      <td>Tryptamine:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tryptophan metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + hista_c + o2_c --&gt; h2o2_c + im4act_c + nh4_c</td>\n",
       "      <td>HISTASE</td>\n",
       "      <td>HISTASE</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + o2_c + ptrc_c --&gt; 4abutn_c + h2o2_c + nh4_c</td>\n",
       "      <td>PTRCOX1</td>\n",
       "      <td>Putrescine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>314</td>\n",
       "      <td>h2o_c + mhista_c + o2_c --&gt; 3mldz_c + h2o2_c + nh4_c</td>\n",
       "      <td>MHISOR</td>\n",
       "      <td>N-Methylhistamine:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>314</td>\n",
       "      <td>13dampp_c + h2o_c + o2_c --&gt; bamppald_c + h2o2_c + nh4_c</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>1,3-Diaminopropane:oxygen oxidoreductase (deaminating)</td>\n",
       "      <td>Beta-Alanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>26</td>\n",
       "      <td>CE1935_e + h2o_e + o2_e --&gt; CE1936_e + h2o2_e + nh4_e</td>\n",
       "      <td>RE0827E</td>\n",
       "      <td>RE0827</td>\n",
       "      <td>Urea cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>26</td>\n",
       "      <td>CE1939_e + h2o_e + o2_e --&gt; CE1943_e + h2o2_e + nh4_e</td>\n",
       "      <td>RE3367E</td>\n",
       "      <td>RE3367</td>\n",
       "      <td>Urea cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>dopa_c + h2o_c + o2_c --&gt; 34dhpac_c + h2o2_c + nh4_c</td>\n",
       "      <td>42A12BOOX</td>\n",
       "      <td>4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>26</td>\n",
       "      <td>h2o_e + o2_e + spmd_e --&gt; CE1939_e + h2o2_e + nh4_e</td>\n",
       "      <td>RE0828E</td>\n",
       "      <td>RE0828</td>\n",
       "      <td>Urea cycle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>26</td>\n",
       "      <td>h2o_c + o2_c + peamn_c --&gt; h2o2_c + nh4_c + pacald_c</td>\n",
       "      <td>PEAMNO</td>\n",
       "      <td>Phenethylamine oxidase</td>\n",
       "      <td>Phenylalanine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>26</td>\n",
       "      <td>h2o_c + mma_c + o2_c --&gt; fald_c + h2o2_c + nh4_c</td>\n",
       "      <td>MAOX</td>\n",
       "      <td>Methylamine:oxygen oxidoreductase (deaminating) (copper-containing)</td>\n",
       "      <td>Tyrosine metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>26</td>\n",
       "      <td>h2o_c + o2_c + trypta_c --&gt; h2o2_c + id3acald_c + nh4_c</td>\n",
       "      <td>TRYPTAOX</td>\n",
       "      <td>Tryptamine:oxygen oxidoreductase(deaminating)(flavin-containing)</td>\n",
       "      <td>Tryptophan metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14457</th>\n",
       "      <td>1351</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14458</th>\n",
       "      <td>1351</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14459</th>\n",
       "      <td>84701</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14460</th>\n",
       "      <td>1347</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14461</th>\n",
       "      <td>1347</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14462</th>\n",
       "      <td>1340</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14463</th>\n",
       "      <td>1340</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14464</th>\n",
       "      <td>4512</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14465</th>\n",
       "      <td>1349</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14466</th>\n",
       "      <td>1349</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14467</th>\n",
       "      <td>1339</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14468</th>\n",
       "      <td>1339</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14469</th>\n",
       "      <td>125965</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14470</th>\n",
       "      <td>1337</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14471</th>\n",
       "      <td>1337</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14472</th>\n",
       "      <td>9377</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14473</th>\n",
       "      <td>9377</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14474</th>\n",
       "      <td>4513</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14475</th>\n",
       "      <td>1329</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14476</th>\n",
       "      <td>1329</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14477</th>\n",
       "      <td>1346</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14478</th>\n",
       "      <td>170712</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14479</th>\n",
       "      <td>170712</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14480</th>\n",
       "      <td>9167</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14481</th>\n",
       "      <td>1345</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14482</th>\n",
       "      <td>1345</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14483</th>\n",
       "      <td>1350</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14484</th>\n",
       "      <td>1350</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14485</th>\n",
       "      <td>1327</td>\n",
       "      <td>4.0 focytC_m + 7.92 h_m + o2_m --&gt; 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>CYOOm3i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14486</th>\n",
       "      <td>1327</td>\n",
       "      <td>4.0 focytC_m + 8.0 h_m + o2_m --&gt; 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>CYOOm2i</td>\n",
       "      <td>Oxidative phosphorylation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14487 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  \\\n",
       "0      0      \n",
       "1      8639   \n",
       "2      8639   \n",
       "3      8639   \n",
       "4      8639   \n",
       "...     ...   \n",
       "14482  1345   \n",
       "14483  1350   \n",
       "14484  1350   \n",
       "14485  1327   \n",
       "14486  1327   \n",
       "\n",
       "                                                                                       rxn  \\\n",
       "0      hdca24g_c <=> hdca24g_r                                                               \n",
       "1      h2o_c + mma_c + o2_c --> fald_c + h2o2_c + nh4_c                                      \n",
       "2      aact_c + h2o_c + o2_c --> h2o2_c + mthgxl_c + nh4_c                                   \n",
       "3      h2o_c + o2_c + ptrc_c --> 4abutn_c + h2o2_c + nh4_c                                   \n",
       "4      dopa_c + h2o_c + o2_c --> 34dhpac_c + h2o2_c + nh4_c                                  \n",
       "...                                                     ...                                  \n",
       "14482  4.0 focytC_m + 8.0 h_m + o2_m --> 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i                  \n",
       "14483  4.0 focytC_m + 7.92 h_m + o2_m --> 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m   \n",
       "14484  4.0 focytC_m + 8.0 h_m + o2_m --> 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i                  \n",
       "14485  4.0 focytC_m + 7.92 h_m + o2_m --> 4.0 ficytC_m + 1.96 h2o_m + 4.0 h_i + 0.02 o2s_m   \n",
       "14486  4.0 focytC_m + 8.0 h_m + o2_m --> 4.0 ficytC_m + 2.0 h2o_m + 4.0 h_i                  \n",
       "\n",
       "          rxn_id  \\\n",
       "0      HDCA24Gtr   \n",
       "1      MAOX        \n",
       "2      AACTOOR     \n",
       "3      PTRCOX1     \n",
       "4      42A12BOOX   \n",
       "...          ...   \n",
       "14482  CYOOm2i     \n",
       "14483  CYOOm3i     \n",
       "14484  CYOOm2i     \n",
       "14485  CYOOm3i     \n",
       "14486  CYOOm2i     \n",
       "\n",
       "                                                                                     rxn_name  \\\n",
       "0      Glucuronidated Compound Transport (HDCA-24G)                                             \n",
       "1      Methylamine:oxygen oxidoreductase (deaminating) (copper-containing)                      \n",
       "2      Aminoacetone:oxygen oxidoreductase(deaminating)(flavin-containing)                       \n",
       "3      Putrescine:oxygen oxidoreductase (deaminating)                                           \n",
       "4      4-(2-Aminoethyl)-1,2-benzenediol:oxygen oxidoreductase(deaminating)(flavin-containing)   \n",
       "...                                                                                       ...   \n",
       "14482  CYOOm2i                                                                                  \n",
       "14483  CYOOm3i                                                                                  \n",
       "14484  CYOOm2i                                                                                  \n",
       "14485  CYOOm3i                                                                                  \n",
       "14486  CYOOm2i                                                                                  \n",
       "\n",
       "                                                subsystem  \n",
       "0      Transport, extracellular                            \n",
       "1      Tyrosine metabolism                                 \n",
       "2      Glycine, serine, alanine, and threonine metabolism  \n",
       "3      Arginine and proline metabolism                     \n",
       "4      Tyrosine metabolism                                 \n",
       "...                    ...                                 \n",
       "14482  Oxidative phosphorylation                           \n",
       "14483  Oxidative phosphorylation                           \n",
       "14484  Oxidative phosphorylation                           \n",
       "14485  Oxidative phosphorylation                           \n",
       "14486  Oxidative phosphorylation                           \n",
       "\n",
       "[14487 rows x 5 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_g_rxn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "sol = model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_go.loc[:,'in_M'] = df_go.gene.isin(genes_sig_M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_go_rxn = pd.merge(df_go, df_g_rxn, on='gene', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "      <th>symbol</th>\n",
       "      <th>rxn</th>\n",
       "      <th>rxn_id</th>\n",
       "      <th>rxn_name</th>\n",
       "      <th>subsystem</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1101</th>\n",
       "      <td>bile acid transmembrane transporter activity</td>\n",
       "      <td>0.00628</td>\n",
       "      <td>SLCO1B3</td>\n",
       "      <td>hco3_c + lca3s_e &lt;=&gt; hco3_e + lca3s_c</td>\n",
       "      <td>LCA3St</td>\n",
       "      <td>Lithocholic acid 3-sulfate transport via bicarbonate countertransport</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1102</th>\n",
       "      <td>bile acid transmembrane transporter activity</td>\n",
       "      <td>0.00628</td>\n",
       "      <td>SLCO1B3</td>\n",
       "      <td>dca3s_e + hco3_c &lt;=&gt; dca3s_c + hco3_e</td>\n",
       "      <td>DCA3St</td>\n",
       "      <td>Deoxycholic acid 3-sulfate transport via bicarbonate countertransport</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103</th>\n",
       "      <td>bile acid transmembrane transporter activity</td>\n",
       "      <td>0.00628</td>\n",
       "      <td>SLCO1B3</td>\n",
       "      <td>7dhcdchol_e + hco3_c &lt;=&gt; 7dhcdchol_c + hco3_e</td>\n",
       "      <td>7DHCDCHOLt</td>\n",
       "      <td>7-Dehydrochenodeoxycholic acid transport via bicarbonate countertransport</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104</th>\n",
       "      <td>bile acid transmembrane transporter activity</td>\n",
       "      <td>0.00628</td>\n",
       "      <td>SLCO1B3</td>\n",
       "      <td>3dhchol_e + hco3_c &lt;=&gt; 3dhchol_c + hco3_e</td>\n",
       "      <td>3DHCHOLt</td>\n",
       "      <td>3-Dehydrocholic acid  transport via bicarbonate countertransport</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>bile acid transmembrane transporter activity</td>\n",
       "      <td>0.00628</td>\n",
       "      <td>SLCO1B3</td>\n",
       "      <td>ca3s_e + hco3_c &lt;=&gt; ca3s_c + hco3_e</td>\n",
       "      <td>CA3St</td>\n",
       "      <td>Cholic acid 3-sulfate transport via bicarbonate countertransport</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              name  p_fdr_bh   symbol  \\\n",
       "1101  bile acid transmembrane transporter activity  0.00628   SLCO1B3   \n",
       "1102  bile acid transmembrane transporter activity  0.00628   SLCO1B3   \n",
       "1103  bile acid transmembrane transporter activity  0.00628   SLCO1B3   \n",
       "1104  bile acid transmembrane transporter activity  0.00628   SLCO1B3   \n",
       "1105  bile acid transmembrane transporter activity  0.00628   SLCO1B3   \n",
       "\n",
       "                                                rxn      rxn_id  \\\n",
       "1101  hco3_c + lca3s_e <=> hco3_e + lca3s_c          LCA3St       \n",
       "1102  dca3s_e + hco3_c <=> dca3s_c + hco3_e          DCA3St       \n",
       "1103  7dhcdchol_e + hco3_c <=> 7dhcdchol_c + hco3_e  7DHCDCHOLt   \n",
       "1104  3dhchol_e + hco3_c <=> 3dhchol_c + hco3_e      3DHCHOLt     \n",
       "1105  ca3s_e + hco3_c <=> ca3s_c + hco3_e            CA3St        \n",
       "\n",
       "                                                                       rxn_name  \\\n",
       "1101  Lithocholic acid 3-sulfate transport via bicarbonate countertransport       \n",
       "1102  Deoxycholic acid 3-sulfate transport via bicarbonate countertransport       \n",
       "1103  7-Dehydrochenodeoxycholic acid transport via bicarbonate countertransport   \n",
       "1104  3-Dehydrocholic acid  transport via bicarbonate countertransport            \n",
       "1105  Cholic acid 3-sulfate transport via bicarbonate countertransport            \n",
       "\n",
       "                     subsystem  \n",
       "1101  Transport, extracellular  \n",
       "1102  Transport, extracellular  \n",
       "1103  Transport, extracellular  \n",
       "1104  Transport, extracellular  \n",
       "1105  Transport, extracellular  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_mol = df_go_rxn[df_go_rxn.namespace.str.contains('molecular')].sort_values(['p_fdr_bh','namespace','name'])\n",
    "cols = ['name','p_fdr_bh','symbol','rxn','rxn_id','rxn_name','subsystem']\n",
    "df_go_mol[df_go_mol.in_M][cols].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p_fdr_bh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>subsystem</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Drug metabolism</th>\n",
       "      <td>0.006849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Transport, extracellular</th>\n",
       "      <td>0.006849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bile acid synthesis</th>\n",
       "      <td>0.037601</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          p_fdr_bh\n",
       "subsystem                         \n",
       "Drug metabolism           0.006849\n",
       "Transport, extracellular  0.006849\n",
       "Bile acid synthesis       0.037601"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols2 = ['p_fdr_bh','subsystem']\n",
    "df_go_mol[df_go_mol.in_M][cols2].drop_duplicates().groupby('subsystem').median().sort_values('p_fdr_bh')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Literature validation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-  \"alterations in FA metabolism could play a pivotal role in production of arrhythmias\":\n",
    "https://www.fasebj.org/doi/abs/10.1096/fasebj.31.1_supplement.782.14\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Transport, extracellular', 'Drug metabolism',\n",
       "       'Bile acid synthesis'], dtype=object)"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_mol.subsystem.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>id</th>\n",
       "      <th>p_fdr_bh</th>\n",
       "      <th>gene</th>\n",
       "      <th>symbol</th>\n",
       "      <th>in_M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [name, namespace, id, p_fdr_bh, gene, symbol, in_M]\n",
       "Index: []"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go[df_go.name.str.contains('channel')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Last step: show using simulation the mechanism by which FA oxidation (and others) cause the Arrhythmia side effect\n",
    "\n",
    "Steps:\n",
    "1. make heart-specific model\n",
    "1. simulate healthy heart metabolism (linked to electrical conduction)\n",
    "1. simulate arrhythmia\n",
    "    1. by perturbing the PoSe-Path genes. Hypothesis: perturbing these genes causes arrhythmia, or that arrhythmia involves change in the reactions encoded by these genes\n",
    "    1. by using known arrhythmia simulation parameters, and check consistency of known arrhythmia mechanisms with the genes found by PoSe-Path [X: DON'T already know arrhythmia parameters]\n",
    "1. Go with A.\n",
    "    - Validate using another metabolic marker of arrhythmia--show that it changes in the correct direction\n",
    "    - Validate using omics (RNA-Seq, proteomics, metabolomics) data from arrhythmia patients\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sol"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "geneid2info = {v.GeneID: (v.Symbol,v.description) for k,v in GeneID2nt_hum.items()}\n",
    "\n",
    "[geneid2info[g] for g in geneids_study if g in geneid2info]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "gg = goea_results_sig[0]\n",
    "gg.GO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://github.com/tanghaibao/goatools/blob/master/notebooks/goea_nbt3102.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "from goatools.godag_plot import plot_gos, plot_results, plot_goid2goobj\n",
    "\n",
    "\n",
    "# Plot subset starting from these significant GO terms\n",
    "goid_subset = [g.GO for g in goea_results_sig] # [\n",
    "#     'GO:0003723', # MF D04 RNA binding (32 genes)\n",
    "#     'GO:0044822', # MF D05 poly(A) RNA binding (86 genes)\n",
    "#     'GO:0003729', # MF D06 mRNA binding (11 genes)\n",
    "#     'GO:0019843', # MF D05 rRNA binding (6 genes)\n",
    "#     'GO:0003746', # MF D06 translation elongation factor activity (5 genes)\n",
    "# ]\n",
    "plot_gos(\"results3/out/go_enrich.png\", \n",
    "    goid_subset, # Source GO ids\n",
    "    obodag, \n",
    "    goea_results=goea_results_all) # Use pvals for coloring\n",
    "# plot_results(\"nbt3102_BP.png\", goea_results_sig)\n",
    "\n",
    "This plot contains GOEA results:\n",
    "\n",
    "- GO terms colored by P-value:\n",
    "    - pval < 0.005 (light red)\n",
    "    - pval < 0.01 (light orange)\n",
    "    - pval < 0.05 (yellow)\n",
    "    - pval > 0.05 (grey) Study terms that are not statistically significant\n",
    "- GO terms with study gene counts printed. e.g., \"32 genes\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "plot_gos(\"results3/out/go_enrich_symbols.pdf\", \n",
    "    goid_subset, # Source GO ids\n",
    "    obodag,\n",
    "    goea_results=goea_results_all, # use pvals for coloring\n",
    "    # We can further configure the plot...\n",
    "    id2symbol=geneid2symbol, # Print study gene Symbols, not Entrez GeneIDs\n",
    "    study_items=6, # Only only 6 gene Symbols max on GO terms\n",
    "    items_p_line=3, # Print 3 genes per line\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
